---
title: "rankEffects_DM_modeling"
author: "c ryan"
date: "1/12/2021"
output: html_document
---

load helpful functions
```{r}
perm.fdr=function(input_df,perm_df,Pvals_col_name,name){
  
  pvals_index=which(colnames(input_df)==Pvals_col_name)
  ro<-input_df[order(input_df[,pvals_index]),]
  p_obs <- data.frame(pvalue=ro[,pvals_index])
  p_vector<-matrix(as.matrix(perm_df),ncol=1)
  p_vector=data.frame(p_vector[order(p_vector)])
  
  F<-p_obs[,1]
  F_o<-p_obs[,1]
  pi_hat<-p_obs[,1]
  
  j=1
  observed<-length(p_obs[,1])
  randoms<-length(p_vector[,1])
  
  for(i in 1:observed)
  {
    repeat
    {
      if((p_vector[j,1]<p_obs[i,1])&j<randoms){j<-j+1}else{break}
    }
    F[i]=i/observed
    F_o[i]=(j-1)/randoms
    if(F_o[i]<1){pi_hat[i]=(1-F[i])/(1-F_o[i])}else{pi_hat[i]=1}
  }
  tabla <-data.frame(pi_hat,pval=p_obs[,1])
  
  tabla[1,]=c(1,0)
  last_percentile_average=mean(tabla$pi_hat[as.integer(min((length(tabla[,1])*0.99),(nrow(tabla)-1)):length(tabla[,1]))])
  tabla[nrow(tabla),]=c(last_percentile_average,1)
  constraint_matrix=as.matrix(data.frame(c(0,2),c(0,1),c(1,0)))
  f_hat<-suppressWarnings(cobs(tabla$pval,tabla$pi_hat,constraint="convex",pointwise=constraint_matrix,maxiter=1000,print.warn=FALSE,print.mesg=FALSE))
  
  f_hat_serie=f_hat$fitted
  pi_o=f_hat_serie[length(f_hat_serie)]
  pi_o=min(pi_o,1)
  pi_o=max(pi_o,0)
  
  Fdr_ST_perm=pi_o*F_o/F
  
  for(i in 1:length(p_obs[,1]))
  {
    Fdr_ST_perm[i]=pi_o*F_o[i]/F[i]
    if(i>1)
    {
      for(j in 1:(i-1))
      {
        if(Fdr_ST_perm[i-j]>Fdr_ST_perm[i]){Fdr_ST_perm[i-j]=Fdr_ST_perm[i]}else{break}
      }
    }
    if(Fdr_ST_perm[i]>1)  Fdr_ST_perm[i]=1
  }
  
  fdrs_df <-data.frame(ro,q_ST_perm=Fdr_ST_perm)
  rownames(fdrs_df)=rownames(ro)
  colnames(fdrs_df)[ncol(fdrs_df)]=paste0("fdr_",name)
  
  return(fdrs_df)
}


plotModelHists <- function(emmremlDF, folderName = "~/Downloads/", modelName = "model") {
  #emmremlDF <- res_fullDF
  #folderName <- figDir
  
  nVars <- dim(emmremlDF)[2] / 3
  
  colNums <- (2 * nVars + 1) : (3 * nVars)
  
  for (c in colNums) {
    ggplot() +
      geom_histogram(aes(x = emmremlDF[,c]), bins = 100) +
      xlab(colnames(emmremlDF)[c]) + ggtitle(paste0("emmreml run ",modelName))
    ggsave(paste0(folderName,"emmreml_hist_",modelName,"_",str_replace(colnames(emmremlDF)[c],":","_"),".png"), width = 6, height = 4)
  }
}

```

```{r data, include=FALSE}
library(ggplot2)
library(limma)
library(edgeR)
library(EMMREML)
library(cobs)
library(cowplot)
library(snakecase)
library(stringr)
library(qvalue)
library(scales)

figDir <- "~/Dropbox (Personal)/scRNA/SGE_DM/figures/"

## read in data frames with DM, mean, and variance for each gene ; these files are in SGE_DM/dataFiles ## 
sge_LPS_DM_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_DM_matrix')
sge_LPS_mean_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_mean_matrix')
sge_LPS_var_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_var_matrix')
sge_LPS_pbMean_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_pbMean_matrix')
sge_NC_DM_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_DM_matrix')
sge_NC_mean_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_mean_matrix')
sge_NC_var_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_var_matrix')
sge_NC_pbMean_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_pbMean_matrix')

## remove genes w/0 as DM values for every individual
## don't have to remove them from 
sge_LPS_DM_matrix <- sge_LPS_DM_matrix[rowSums(sge_LPS_DM_matrix) != 0,]
sge_NC_DM_matrix <- sge_NC_DM_matrix[rowSums(sge_NC_DM_matrix) != 0,]

## remove chrMT counts
#sge_LPS_DM_matrix <- sge_LPS_DM_matrix[rownames(sge_LPS_DM_matrix) != "chrMT",]
#sge_NC_DM_matrix <- sge_NC_DM_matrix[rownames(sge_NC_DM_matrix) != "chrMT",]

## only use common set of genes - must exist in all matrices (8)
geneList <- rownames(table(c(rownames(sge_NC_DM_matrix),
            rownames(sge_NC_var_matrix),
            rownames(sge_NC_mean_matrix),
            rownames(sge_NC_pbMean_matrix),
            rownames(sge_LPS_DM_matrix),
            rownames(sge_LPS_var_matrix),
            rownames(sge_LPS_mean_matrix),
            rownames(sge_LPS_pbMean_matrix))))[table(c(rownames(sge_NC_DM_matrix),
            rownames(sge_NC_var_matrix),
            rownames(sge_NC_mean_matrix),
            rownames(sge_NC_pbMean_matrix),
            rownames(sge_LPS_DM_matrix),
            rownames(sge_LPS_var_matrix),
            rownames(sge_LPS_mean_matrix),
            rownames(sge_LPS_pbMean_matrix))) == 8]

geneList <- geneList[geneList != "chrMT"]

## subset mean and variance data frames to include only the genes present in the DM data frame ##
sge_NC_DM_matrix <- subset(sge_NC_DM_matrix, rownames(sge_NC_DM_matrix) %in% geneList)
sge_NC_mean_matrix <- subset(sge_NC_mean_matrix, rownames(sge_NC_mean_matrix) %in% geneList)
sge_NC_var_matrix <- subset(sge_NC_var_matrix, rownames(sge_NC_var_matrix) %in% geneList)
sge_NC_pbMean_matrix <- subset(sge_NC_pbMean_matrix, rownames(sge_NC_pbMean_matrix) %in% geneList)

sge_LPS_DM_matrix <- subset(sge_LPS_DM_matrix, rownames(sge_LPS_DM_matrix) %in% geneList)
sge_LPS_mean_matrix <- subset(sge_LPS_mean_matrix, rownames(sge_LPS_mean_matrix) %in% geneList)
sge_LPS_var_matrix <- subset(sge_LPS_var_matrix, rownames(sge_LPS_var_matrix) %in% geneList)
sge_LPS_pbMean_matrix <- subset(sge_LPS_pbMean_matrix, rownames(sge_LPS_pbMean_matrix) %in% geneList)


```


Does the DM metric adequately remove the mean/variance relationship?
Plot mean vs variance & mean v DM for a few individuals to observe initial pattern
```{r}
p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 1') +
  ylab('variance in gene expression - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 2') +
  ylab('variance in gene expression - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 3') +
  ylab('variance in gene expression - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_mean_ge__by_individual_noLim.png"))

p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  #ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 1') +
  ylab('variance in gene expression - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  #ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 2') +
  ylab('variance in gene expression - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  #ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 3') +
  ylab('variance in gene expression - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_mean_ge__by_individual_yLim250.png"))

p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_DM_matrix[,1])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 1') +
  ylab('DM - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_DM_matrix[,2])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 2') +
  ylab('DM - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_DM_matrix[,3])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 3') +
  ylab('DM - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"DM_v_mean_ge__by_individual.png"))
```

how about pesudobulk mean vs DM and variance for same individuals
```{r}
p1 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,1],sge_NC_DM_matrix[,1])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('pbMean ge - individual 1') +
  ylab('DM - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,2],sge_NC_DM_matrix[,2])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('pbMean ge - individual 2') +
  ylab('DM - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,3],sge_NC_DM_matrix[,3])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('pbMean ge - individual 3') +
  ylab('DM - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"DM_v_pbMean_ge__by_individual.png"))



p1 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('pbMean ge - individual 1') +
  ylab('variance in ge - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('pbMean ge - individual 2') +
  ylab('variance in ge - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('pbMean ge - individual 3') +
  ylab('variance in ge - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_pbMean_ge__by_individual.png"))
```


make mean plot more readable with log
```{r}
# + scale_x_continuous(trans = "log2")

p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 1') +
  ylab('variance in gene expression - individual 1') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 2') +
  ylab('variance in gene expression - individual 2') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 3') +
  ylab('variance in gene expression - individual 3') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_log2_mean_ge__by_individual.png"))

p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_DM_matrix[,1])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 1') +
  ylab('DM - individual 1') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_DM_matrix[,2])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 2') +
  ylab('DM - individual 2') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_DM_matrix[,3])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 3') +
  ylab('DM - individual 3') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"DM_v_log2_mean_ge__by_individual.png"))


```


```{r}
## just take a quick look at some stuff ##
meanVarRs <- meanDMRs <- pbMeanVarRs <- pbMeanDMRs <- vector()

for (i in 1:45){
  meanVarRs[[i]] <- cor.test(sge_NC_mean_matrix[,i],sge_NC_var_matrix[,i], method = "spearman")$estimate 
}

for (i in 1:45){
  meanDMRs[[i]] <- cor.test(sge_NC_mean_matrix[,i],sge_NC_DM_matrix[,i], method = "spearman")$estimate 
}

for (i in 1:45){
  pbMeanVarRs[[i]] <- cor.test(sge_NC_pbMean_matrix[,i],sge_NC_var_matrix[,i], method = "spearman")$estimate 
}

for (i in 1:45){
  pbMeanDMRs[[i]] <- cor.test(sge_NC_pbMean_matrix[,i],sge_NC_DM_matrix[,i], method = "spearman")$estimate 
}


ggplot() +
  geom_density(aes(meanVarRs, fill = 'mean/variance')) +
  geom_density(aes(meanDMRs, fill = 'mean/DM')) +
  #xlim(-0.01,1) +
  xlab('R') +
  scale_fill_discrete(name = "")
ggsave(paste0(figDir,"hist_mean_dm_and_var_Rvalues.png"))

ggplot() +
  geom_density(aes(pbMeanVarRs, fill = 'pbMean/variance')) +
  geom_density(aes(pbMeanDMRs, fill = 'pbMean/DM')) +
  #xlim(-0.01,1) +
  xlab('R') +
  scale_fill_discrete(name = "")
ggsave(paste0(figDir,"hist_pbMean_dm_and_var_Rvalues.png"))

ggplot() +
  geom_density(aes(meanVarRs, fill = 'mean/variance')) +
  geom_density(aes(pbMeanVarRs, fill = 'pbMean/variance')) +
  #xlim(-0.01,1) +
  xlab('R') +
  scale_fill_discrete(name = "")
ggsave(paste0(figDir,"hist_means_and_var_Rvalues.png"))

ggplot() +
  geom_density(aes(meanDMRs, fill = 'mean/DM')) +
  geom_density(aes(pbMeanDMRs, fill = 'pbMean/DM')) +
  #xlim(-0.01,1) +
  xlab('R') +
  scale_fill_discrete(name = "")
ggsave(paste0(figDir,"hist_means_and_dm_Rvalues.png"))



ggplot() +
  geom_density(aes(abs(sge_NC_DM_matrix[,1]), fill='NC'), alpha=0.5) +
  geom_density(aes(abs(sge_LPS_DM_matrix[,1]), fill='LPS'), alpha=0.5) +
  xlim(0,0.4)

```

Here I'm doing some modeling to ask what are the effects of treatment and rank on DM & get some metadata

```{r}
## Here I'm doing some modeling to ask what are the effects of treatment and rank on DM ##
## get some metadata ##
metadata <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_metadata_matrix', header = T) # this files is also in SGE_DM/dataFiles
metadata[1,1] <- 'Mo'
metadata <- metadata[order(metadata$ID),]

## make sure metadata order matches data frames ##
all.equal(metadata$ID,colnames(sge_NC_DM_matrix))

## double the metadata so it includes both NC and LPS ##
metadata <- rbind(metadata,metadata)
metadata$trt <- rep(c('NC','LPS'), times=c(45,45))

## generate a dataframe of NC and LPS combined ##
sge_DM_matrix_nested <- merge(sge_NC_DM_matrix,sge_LPS_DM_matrix, by=0)
rownames(sge_DM_matrix_nested) <- sge_DM_matrix_nested$Row.names
sge_DM_matrix_nested$Row.names <- NULL
```

Is controlling for group necessary? (looks like yes)
```{r}
#no control for group
pcaNoLM <- prcomp(cor(sge_DM_matrix_nested, method = "spearman"))

df_out <- as.data.frame(pcaNoLM$x)
dfMetaNoLM <- cbind(df_out,metadata)

#pcaNoLM <- pca

summary(pcaNoLM)

ggplot(dfMetaNoLM,aes(x=PC1,y=PC2,color=group)) + geom_point(size=3) + 
   xlab(paste0("PC1: ",100 * round(summary(pcaNoLM)$imp[2,1],3),"% variance")) + 
   ylab(paste0("PC2: ",100 * round(summary(pcaNoLM)$imp[2,2],3),"% variance")) +
   ggtitle(paste0("PCA - Individuals (directly on DM cors)")) + 
   theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(figDir,"pca_dm_colGroup_noLM.png"))

ggplot(dfMetaNoLM,aes(x = group, y = PC1, fill = group)) + geom_boxplot() + 
   ylab(paste0("PC1: ",100 * round(summary(pcaNoLM)$imp[2,1],3),"% variance")) + 
   xlab(paste0("Group")) +
   ggtitle(paste0("PCA - Individuals (directly on DM cors)")) + 
   theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(figDir,"pca_dm_PC1_boxpot_noLM.png"))

ggplot(dfMetaNoLM,aes(x = group, y = PC2, fill = group)) + geom_boxplot() + 
   ylab(paste0("PC2: ",100 * round(summary(pcaNoLM)$imp[2,2],3),"% variance")) + 
   xlab(paste0("Group")) +
   ggtitle(paste0("PCA - Individuals (directly on DM cors)")) + 
   theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(figDir,"pca_dm_PC2_boxpot_noLM.png"))

#with control for group
design <- model.matrix(~metadata$group)
fit <-lmFit(sge_DM_matrix_nested,design)
intercepts=data.frame(eBayes(fit))[,1]
resid_DM=apply(residuals.MArrayLM(object=fit, sge_DM_matrix_nested),2,function(x){x+intercepts})
pca <- prcomp(cor(resid_DM, method = "spearman"))

df_out <- as.data.frame(pca$x)
dfMeta <- cbind(df_out,metadata)

#pcaNoLM <- pca

summary(pca)

ggplot(dfMeta,aes(x=PC1,y=PC2,color=group)) + geom_point(size=3) + 
   xlab(paste0("PC1: ",100 * round(summary(pca)$imp[2,1],3),"% variance")) + 
   ylab(paste0("PC2: ",100 * round(summary(pca)$imp[2,2],3),"% variance")) +
   ggtitle(paste0("PCA - Individuals (control for Group)")) + 
   theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(figDir,"pca_dm_colGroup.png"))

ggplot(dfMeta,aes(x = group, y = PC1, fill = group)) + geom_boxplot() + 
   ylab(paste0("PC1: ",100 * round(summary(pca)$imp[2,1],3),"% variance")) + 
   xlab(paste0("Group")) +
   ggtitle(paste0("PCA - Individuals (control for Group)")) + 
   theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(figDir,"pca_dm_PC1_boxpot_LMgroup.png"))

ggplot(dfMeta,aes(x = group, y = PC2, fill = group)) + geom_boxplot() + 
   ylab(paste0("PC2: ",100 * round(summary(pca)$imp[2,2],3),"% variance")) + 
   xlab(paste0("Group")) +
   ggtitle(paste0("PCA - Individuals (control for Group)")) + 
   theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(figDir,"pca_dm_PC2_boxpot_LMgroup.png"))



```


```{r}
## regress out group effects from DM ##
design <- model.matrix(~metadata$group)
fit <-lmFit(sge_DM_matrix_nested,design)
intercepts=data.frame(eBayes(fit))[,1]
resid_DM=apply(residuals.MArrayLM(object=fit, sge_DM_matrix_nested),2,function(x){x+intercepts})

## run a nested model with trt, and age and elo nested within trt ##
## elo and age have to be mean centered for nested model ##
metadata$elo_centered <- scale(metadata$elo, center = T, scale = F)
#metadata$elo_centered <- metadata$elo - mean(metadata$elo)
metadata$age_centered <- scale(metadata$age, center = T, scale = F)
#metadata$age_centered <- metadata$age - mean(metadata$age)

design = model.matrix(~trt+trt:elo_centered+trt:age_centered,data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_DM[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_DM[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}
```






Fit a model for each gene using emmreml
```{r}
for(i in 1:nrow(resid_DM))
{
  emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_fullDF <- data.frame(res_full)
```

```{r direction of LPS effect}

qobj <- qvalue(p = res_fullDF$p_value_trtNC, fdr.level = .1)

res_fullDF_qSig <- res_fullDF[qobj$significant,]

#what is the direction of the genes that pass an FDR < 0.10

ggplot() +
  geom_histogram(aes(x = (res_fullDF$p_value_trtNC)), bins = 100)

ggplot() +
  geom_histogram(aes(x = (qobj$qvalues)), bins = 100) +
  geom_histogram(aes(x = (qobj$pvalues)), alpha = .5, bins = 100)


ggplot() +
  geom_point(aes(x = res_fullDF$beta_trtNC, y = -log10(res_fullDF$p_value_trtNC)))

ggplot() +
  geom_point(aes(x = res_fullDF_qSig$beta_trtNC, y = -log10(res_fullDF_qSig$p_value_trtNC)))

ggplot() +
  geom_histogram(aes(x = res_fullDF_qSig$beta_trtNC), bins = 100) +
  geom_vline(xintercept = mean(res_fullDF_qSig$beta_trtNC)) +
  xlim(c(-1.05,1.05))
```


```{r}

plotModelHists(res_fullDF, folderName = figDir, modelName = "DM_model")

ggplot() +
  geom_histogram(aes(x = res_fullDF$p_value_.Intercept.), bins = 100)
ggplot() +
  geom_histogram(aes(x = res_fullDF$p_value_trtNC), bins = 100)
ggplot() +
  geom_histogram(aes(x = res_fullDF$`p_value_trtNC.elo_centered`), bins = 100)
ggplot() +
  geom_histogram(aes(x = res_fullDF$`p_value_trtLPS.elo_centered`), bins = 100)
ggplot() +
  geom_histogram(aes(x = res_fullDF$`p_value_trtNC.age_centered`), bins = 100)
ggplot() +
  geom_histogram(aes(x = res_fullDF$`p_value_trtLPS.age_centered`), bins = 100)
```


```{r}
## Permutations for emprirical null ##
################################################################################################################################
## Run iters iterations of the model after permutting Elo ratings to retrieve an empiric distribution of p-values for rank effects
################################################################################################################################
metadata$flag=0
for(i in 1:45)
{
  metadata$flag[which(metadata$ID == metadata[1:45,'ID'][i])[1]]=1
}

iters = 3
cols_random<-metadata

for(iter in 1:iters)
{
  print(iter)
  
  #Permute Elo among the set of samples flagged as 1 (one sample per individual)
  cols_random$elo[which(cols_random$flag==1)]=sample(cols_random$elo[which(cols_random$flag==1)])
  
  #Complete the permuted Elos of the other samples (this way the mapping individual-to-rank is conserved in the permutations)
  for(i in 1:45)
  {
    set=which(cols_random$ID==metadata[1:45,'ID'][i] & cols_random$flag==0)
    set_ref=which(cols_random$ID==metadata[1:45,'ID'][i] & cols_random$flag==1)
    if(length(set)==1)
      cols_random$elo[set]=cols_random$elo[set_ref]
  }	
  
  #Declare null (i.e. based on permutations) nested design for fixed effects.
  design = model.matrix(~trt+trt:elo_centered+trt:age_centered,data=cols_random)
  
  #Declare object res_null to store in it the permutations' p-values:
  res_null=resid_DM[,1:(ncol(design))]
  colnames(res_null)[1:ncol(design)]=paste0("p_value_",colnames(design))
  
  #Fit a model for each gene using emmreml
  for(i in 1:nrow(resid_DM))
  {
    emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
    res_null[i,]=t(c(emma$pvalbeta[,"none"]))
  }
  
  #we register p-values of the associations to Elo at NC and LPS alone.
  if(iter==1)
  {
    shuffled_elos_pvals_NC <-data.frame(x=res_null[,"p_value_trtNC:elo_centered"])
    shuffled_elos_pvals_LPS <-data.frame(x=res_null[,"p_value_trtLPS:elo_centered"])
    
    rownames(shuffled_elos_pvals_NC)=rownames(res_null)
    rownames(shuffled_elos_pvals_LPS)=rownames(res_null)
  } else {
    shuffled_elos_pvals_NC <- cbind(shuffled_elos_pvals_NC,x=res_null[,"p_value_trtNC:elo_centered"])
    shuffled_elos_pvals_LPS <- cbind(shuffled_elos_pvals_LPS,x=res_null[,"p_value_trtLPS:elo_centered"])
  }
}

```


```{r}
###################################################################################################
## Run iters iterations of the model after permuting Condition to retrieve an empirical distribution
## of p-values for LPS stimulation effect (at average rank, age and tissue composition)
###################################################################################################

cols_random<-metadata

for(iter in 1:iters)
{
  print(iter)
  
  #Permute Elo among the set of samples flagged as 1 (one sample per individual)
  cols_random$trt  <- sample(cols_random$trt)
  
  #Declare null (i.e. based on permutations) nested design for fixed effects.
  design = model.matrix(~trt+trt:elo_centered+trt:age_centered,data=cols_random)
  
  #Declare object res_null to store in it the permutations' p-values:
  res_null=resid_DM[,1:(ncol(design))]
  colnames(res_null)[1:ncol(design)]=paste0("p_value_",colnames(design))
  
  #Fit a model for each gene using emmreml
  for(i in 1:nrow(resid_DM))
  {
    emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
    res_null[i,]=t(c(emma$pvalbeta[,"none"]))
  }
  
  #we register p-values of the associations to Elo at NC and LPS alone.
  if(iter==1)
  {
    shuffled_elos_pvals_LPS_effect <-data.frame(x=res_null[,"p_value_trtNC"])
    rownames(shuffled_elos_pvals_LPS_effect)=rownames(res_null)
  } else {
    shuffled_elos_pvals_LPS_effect <- cbind(shuffled_elos_pvals_LPS_effect,x=res_null[,"p_value_trtNC"])
  }
}
```





```{r}
### Ryan, this is the part that is breaking, let me know if it works for you ###



res_fullDF <- perm.fdr(res_fullDF,shuffled_elos_pvals_NC,"p_value_trtNC.elo_centered","eloNC")

res_fullDF <- perm.fdr(res_fullDF,shuffled_elos_pvals_LPS,"p_value_trtLPS.elo_centered","eloLPS")

res_fullDF <- perm.fdr(res_fullDF,shuffled_elos_pvals_LPS_effect,"p_value_trtNC","test")
```





how do those plots change if we cut off the bottom X% of genes?
```{r pvalue distribution without low GE genes}
percents <- c(5,10,15,25,50,75)

for (p in percents) {
  cutoff <- quantile(rowMeans(sge_mean_matrix_nested), c(p/100))
  
  plotResFull <- res_fullDF[rowMeans(sge_mean_matrix_nested) > cutoff,]
  
  ggplot() +
    geom_histogram(aes(x = res_fullDF$p_value_trtLPS.elo_centered), alpha = .5, bins = 100) +
    geom_histogram(aes(x = plotResFull$p_value_trtLPS.elo_centered), bins = 100) +
    ggtitle(paste0("LPS:elo_centered p-value dist | bottom ",p,"% of genes removed"))
  ggsave(paste0(figDir,"DM_LPSelo_pvalue_",p,"per_genes_removed.png"))
}


p <- 50
cutoff <- quantile(rowMeans(sge_mean_matrix_nested), c(p/100))

plotResFull <- res_fullDF[rowMeans(sge_mean_matrix_nested) < cutoff,]

ggplot() +
  geom_histogram(aes(x = res_fullDF$p_value_trtLPS.elo_centered), alpha = .5, bins = 100, fill = "red") +
  geom_histogram(aes(x = plotResFull$p_value_trtLPS.elo_centered), bins = 100, fill = "red") +
  ggtitle(paste0("LPS:elo_centered p-value dist | bottom ",p,"% of genes KEPT"))
ggsave(paste0(figDir,"DM_LPSelo_pvalue_",p,"per_TOP_genes_kept.png"))
```



what if we use variance but control for mean?
```{r model variance controlling for mean}
## Here I'm doing some modeling to ask what are the effects of treatment and rank on variance ##
## get some metadata ##
metadata <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_metadata_matrix', header = T) # this files is also in SGE_DM/dataFiles
metadata[1,1] <- 'Mo'
metadata <- metadata[order(metadata$ID),]

## make sure metadata order matches data frames ##
all.equal(metadata$ID,colnames(sge_NC_var_matrix))

## double the metadata so it includes both NC and LPS ##
metadata <- rbind(metadata,metadata)
metadata$trt <- rep(c('NC','LPS'), times=c(45,45))

## generate a dataframe of NC and LPS combined ##
sge_var_matrix_nested <- merge(sge_NC_var_matrix,sge_LPS_var_matrix, by=0)
rownames(sge_var_matrix_nested) <- sge_var_matrix_nested$Row.names
sge_var_matrix_nested$Row.names <- NULL

##### NEED TO COMBINE MEANS, THEN GET A MEAN FOR EACH ANIMAL, 

sge_mean_matrix_nested <- merge(sge_NC_mean_matrix,sge_LPS_mean_matrix, by=0)
rownames(sge_mean_matrix_nested) <- sge_mean_matrix_nested$Row.names
sge_mean_matrix_nested$Row.names <- NULL

ggplot() +
  geom_histogram(aes(x = rowMeans(sge_mean_matrix_nested)), bins = 50) +
  xlim(c(0,5))

##### AND ADD THAT TO METADATA
metadata$means <- colMeans(sge_mean_matrix_nested)

##### THEN I CAN CONTROL FOR THE MEAN WITH THE VARIANCES AS THE FOCAL VARIABLE IN PLACE OF DM
## regress out group effects from DM ##
design <- model.matrix(~metadata$means + metadata$group)
fit <-lmFit(sge_var_matrix_nested,design)
intercepts=data.frame(eBayes(fit))[,1]
resid_varMean=apply(residuals.MArrayLM(object=fit, sge_var_matrix_nested),2,function(x){x+intercepts})

## run a nested model with trt, and age and elo nested within trt ##
## elo and age have to be mean centered for nested model ##
metadata$elo_centered <- scale(metadata$elo, center = T, scale = F)
#metadata$elo_centered <- metadata$elo - mean(metadata$elo)
metadata$age_centered <- scale(metadata$age, center = T, scale = F)
#metadata$age_centered <- metadata$age - mean(metadata$age)

design = model.matrix(~trt+trt:elo_centered+trt:age_centered,data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_varMean[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_varMean[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}

#emmreml
for(i in 1:nrow(resid_varMean))
{
  emma=emmreml(y=resid_varMean[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_fullvarDF <- as.data.frame(res_full)

plotModelHists(res_fullvarDF, folderName = figDir, modelName = "Var_model_control_mean")

ggplot() +
  geom_histogram(aes(x = res_fullvarDF$p_value_trtNC), bins = 100)
ggplot() +
  geom_histogram(aes(x = res_fullvarDF$`p_value_trtNC:elo_centered`), bins = 100)
ggplot() +
  geom_histogram(aes(x = res_fullvarDF$`p_value_trtLPS:elo_centered`), bins = 100)

ggplot() +
  geom_histogram(aes(x = res_fullvarDF$`p_value_trtNC:age_centered`), bins = 100)
ggplot() +
  geom_histogram(aes(x = res_fullvarDF$`p_value_trtLPS:age_centered`), bins = 100)

```

plot standardized betas of elo models, DM v variance
```{r}

lpsEloR <- cor.test(res_fullvarDF$`beta_trtLPS:elo_centered` / (res_fullvarDF$`sdev_trtLPS:elo_centered`)^0.5,
                    res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5)$estimate

ggplot() +
  geom_point(aes(x = res_fullvarDF$`beta_trtLPS:elo_centered` / (res_fullvarDF$`sdev_trtLPS:elo_centered`)^0.5,
                 y = res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5),
             alpha = .1) +
  xlab("variance model - std betas LPS:elo") +
  ylab("DM model - std betas LPS:elo") +
  geom_smooth(aes(x = res_fullvarDF$`beta_trtLPS:elo_centered` / (res_fullvarDF$`sdev_trtLPS:elo_centered`)^0.5,
                 y = res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5), method = "lm") +
  geom_text(aes(x = .95 * min(res_fullvarDF$`beta_trtLPS:elo_centered` / (res_fullvarDF$`sdev_trtLPS:elo_centered`)^0.5),
                y = .95 * max(res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5)),
                label = paste0("R = ",round(lpsEloR, digits = 3)))
ggsave(paste0(figDir,"stdBeta_comp_LPSelo_DM_v_Var.png"))


trtR <- cor.test(res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5,
                 res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5)$estimate

ggplot() +
  geom_point(aes(x = res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5,
                 y = res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5),
             alpha = .1) +
  xlab("variance model - std betas LPS") +
  ylab("DM model - std betas LPS") +
  geom_smooth(aes(x = res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5,
                 y = res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5), method = "lm") +
  geom_text(aes(x = .95 * min(res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5),
                y = .95 * max(res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5)),
                label = paste0("R = ",round(trtR, digits = 3)))
ggsave(paste0(figDir,"stdBeta_comp_Trt_DM_v_Var.png"))


```
density plot of sig gene betas
```{r}

VARstdBetas <- res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5
DMstdBetas <- res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5

ggplot() +
  geom_histogram(aes(x = VARstdBetas), fill = "red", alpha = .4, bins = 50) +
  geom_histogram(aes(x = DMstdBetas), fill = "blue", alpha = .4, bins = 50) +
  xlab("Treatment Standardized Betas (----> Higher Var in NC)") +
  geom_text(aes(x = -20,
                y = 850,
                label = "Var Model"), col = "red") +
  geom_text(aes(x = -20,
                y = 700,
                label = "DM Model"), col = "blue")

ggsave(paste0(figDir,"stdBeta_hist_Trt_DM_and_Var__sigOnly.png"))

DMstdBetasSIG <- DMstdBetas[qvalue(p = res_fullDF$p_value_trtNC, fdr.level = .1)$significant]
VARstdBetasSIG <- VARstdBetas[qvalue(p = res_fullvarDF$p_value_trtNC, fdr.level = .1)$significant]

ggplot() +
  geom_histogram(aes(x = VARstdBetasSIG), fill = "red", alpha = .4, bins = 50) +
  geom_histogram(aes(x = DMstdBetasSIG), fill = "blue", alpha = .4, bins = 50) +
  xlab("Significant Treatment Standardized Betas (----> Higher Var in NC)") +
  geom_text(aes(x = -20,
                y = 750,
                label = "Var Model"), col = "red") +
  geom_text(aes(x = -20,
                y = 700,
                label = "DM Model"), col = "blue")

ggsave(paste0(figDir,"stdBeta_hist_Trt_DM_and_Var__sigOnly.png"))

```


what if we use variance but control for pbMean?
```{r model variance controlling for pbMean}
## Here I'm doing some modeling to ask what are the effects of treatment and rank on variance ##
## get some metadata ##
metadata <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_metadata_matrix', header = T) # this files is also in SGE_DM/dataFiles
metadata[1,1] <- 'Mo'
metadata <- metadata[order(metadata$ID),]

## make sure metadata order matches data frames ##
all.equal(metadata$ID,colnames(sge_NC_var_matrix))

## double the metadata so it includes both NC and LPS ##
metadata <- rbind(metadata,metadata)
metadata$trt <- rep(c('NC','LPS'), times=c(45,45))

## generate a dataframe of NC and LPS combined ##
sge_var_matrix_nested <- merge(sge_NC_var_matrix,sge_LPS_var_matrix, by=0)
rownames(sge_var_matrix_nested) <- sge_var_matrix_nested$Row.names
sge_var_matrix_nested$Row.names <- NULL

##### NEED TO COMBINE MEANS, THEN GET A MEAN FOR EACH ANIMAL, 
sge_pbMean_matrix_nested <- merge(sge_NC_pbMean_matrix,sge_LPS_pbMean_matrix, by=0)
rownames(sge_pbMean_matrix_nested) <- sge_pbMean_matrix_nested$Row.names
sge_pbMean_matrix_nested$Row.names <- NULL


ggplot() +
  geom_histogram(aes(x = rowMeans(sge_pbMean_matrix_nested)), bins = 50) +
  xlim(c(0,5))

##### AND ADD THAT TO METADATA
metadata$pbMeans <- colMeans(sge_pbMean_matrix_nested)

##### THEN I CAN CONTROL FOR THE MEAN WITH THE VARIANCES AS THE FOCAL VARIABLE IN PLACE OF DM
## regress out group effects from DM ##
design <- model.matrix(~metadata$pbMeans + metadata$group)
fit <-lmFit(sge_var_matrix_nested,design)
intercepts=data.frame(eBayes(fit))[,1]
resid_varPBMean=apply(residuals.MArrayLM(object=fit, sge_var_matrix_nested),2,function(x){x+intercepts})

## run a nested model with trt, and age and elo nested within trt ##
## elo and age have to be mean centered for nested model ##
metadata$elo_centered <- scale(metadata$elo, center = T, scale = F)
#metadata$elo_centered <- metadata$elo - mean(metadata$elo)
metadata$age_centered <- scale(metadata$age, center = T, scale = F)
#metadata$age_centered <- metadata$age - mean(metadata$age)

design = model.matrix(~trt+trt:elo_centered+trt:age_centered,data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_varPBMean[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_varPBMean[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}

#emmreml
for(i in 1:nrow(resid_varPBMean))
{
  emma=emmreml(y=resid_varPBMean[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_fullvarPBmeanDF <- as.data.frame(res_full)

plotModelHists(res_fullvarPBmeanDF, folderName = figDir, modelName = "Var_model_control_pbMean")

ggplot() +
  geom_histogram(aes(x = res_fullvarPBmeanDF$p_value_trtNC), bins = 100)
ggplot() +
  geom_histogram(aes(x = res_fullvarPBmeanDF$`p_value_trtNC:elo_centered`), bins = 100)
ggplot() +
  geom_histogram(aes(x = res_fullvarPBmeanDF$`p_value_trtLPS:elo_centered`), bins = 100)

ggplot() +
  geom_histogram(aes(x = res_fullvarPBmeanDF$`p_value_trtNC:age_centered`), bins = 100)
ggplot() +
  geom_histogram(aes(x = res_fullvarPBmeanDF$`p_value_trtLPS:age_centered`), bins = 100)

```


```{r}
individMeans <- as.vector(t(sge_mean_matrix_nested))

ggplot() +
  geom_histogram(aes(x = rowMeans(sge_mean_matrix_nested)), bins = 100)

ggplot() +
  geom_histogram(aes(x = individMeans), bins = 31) +
  xlim(c(0,30))

sum(individMeans < 2) / length(individMeans)
sum(individMeans < 1) / length(individMeans)
sum(individMeans < .5) / length(individMeans)
```



Does the variance w/means regressed out remove the mean/variance relationship?
Plot mean vs variance_resids for a few individuals to observe initial pattern
```{r plot variance residuals v mean}

p1 <- ggplot(sge_mean_matrix_nested, aes(sge_mean_matrix_nested[,1],resid_varMean[,1])) +
  geom_point(alpha=0.5) +
  #ylim(-500,500) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean gene expression - individual 1') +
  ylab('variance Resid in gene expression - individual 1') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p2 <- ggplot(sge_mean_matrix_nested, aes(sge_mean_matrix_nested[,2],resid_varMean[,2])) +
  geom_point(alpha=0.5) +
  #ylim(-500,500) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean gene expression - individual 2') +
  ylab('variance Resid in gene expression - individual 2') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p3 <- ggplot(sge_mean_matrix_nested, aes(sge_mean_matrix_nested[,3],resid_varMean[,3])) +
  geom_point(alpha=0.5) +
  #ylim(-500,500) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean gene expression - individual 3') +
  ylab('variance Resid in gene expression - individual 3') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"varResidmean_v_mean_ge__by_individual.png"))
```


```{r plot variance residuals pbMean v mean}

p1 <- ggplot(sge_pbMean_matrix_nested, aes(sge_pbMean_matrix_nested[,1],resid_varPBMean[,1])) +
  geom_point(alpha=0.5) +
  #ylim(-500,500) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('pbMean ge - individual 1') +
  ylab('variance Resid pbMean in ge - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_pbMean_matrix_nested, aes(sge_pbMean_matrix_nested[,2],resid_varPBMean[,2])) +
  geom_point(alpha=0.5) +
  #ylim(-500,500) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('pbMean ge - individual 2') +
  ylab('variance Resid pbMean in ge - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_pbMean_matrix_nested, aes(sge_pbMean_matrix_nested[,3],resid_varPBMean[,3])) +
  geom_point(alpha=0.5) +
  #ylim(-500,500) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('pbMean ge - individual 3') +
  ylab('variance Resid pbMean in ge - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"varResidpbMean_v_mean_ge__by_individual.png"))
```

how does the effect of DM look in other behavioral variables
aggDiff
Aggression Received
Overall Grooming

```{r}
behaveData <- read.table("~/Dropbox (Personal)/scRNA/SGE_DM/data/SGEII_METADATA_MASTER.txt", header = T)

#phase 1 only
behaveData <- subset(behaveData, phase == 1)

#animals we need data for
# behave data is 2 character upper, genomic data is 4 char upper/lower
needIDs <- to_any_case(unique(substr(metadata$ID, start = 1, stop  = 2)), case = "screaming_snake")

#limit behavioral data to correct 45 individs
behaveData <- subset(behaveData, ID %in% needIDs)

#sort by ID
behaveData <- behaveData[order(behaveData$ID),]

#match size of metadata object, repeats for NC & LPS
behaveData <- rbind(behaveData,behaveData)

#get the important variables, centered
metadata$aggAsymm_cent <- behaveData$gc.agg.Diff
metadata$aggRec_cent <- behaveData$gc.overall.agg.rec
metadata$groom_cent <- behaveData$gc.overall.groom

## regress out group effects from DM ##
design <- model.matrix(~metadata$group)
fit <-lmFit(sge_DM_matrix_nested,design)
intercepts=data.frame(eBayes(fit))[,1]
resid_DM=apply(residuals.MArrayLM(object=fit, sge_DM_matrix_nested),2,function(x){x+intercepts})
```

aggAsymm
```{r}
design = model.matrix(~trt+trt:aggAsymm_cent + trt:age_centered, data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_DM[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_DM[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}

for(i in 1:nrow(resid_DM))
{
  emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_full_aggAsymDF <- as.data.frame(res_full)

plotModelHists(res_full_aggAsymDF, folderName = figDir, modelName = "DM_model_AgAsym")
```

Aggression Received
```{r}
design = model.matrix(~trt+trt:aggRec_cent + trt:age_centered, data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_DM[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_DM[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}

for(i in 1:nrow(resid_DM))
{
  emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_full_aggRecDF <- as.data.frame(res_full)
plotModelHists(res_full_aggRecDF, folderName = figDir, modelName = "DM_model_AgRec")
```

Overall Grooming
```{r}
design = model.matrix(~trt+trt:groom_cent + trt:age_centered, data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_DM[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_DM[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}

for(i in 1:nrow(resid_DM))
{
  emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_full_groomDF <- as.data.frame(res_full)
plotModelHists(res_full_groomDF, folderName = figDir, modelName = "DM_model_groom")
```

```{r}
ggplot() +
  geom_histogram(aes(x = res_fullDF$p_value_trtLPS.elo_centered), bins = 100)
ggsave(paste0(figDir,"DM_v_mean__LPSelo_pvalue.png"))

ggplot() +
  geom_histogram(aes(x = res_full_aggAsymDF$`p_value_trtLPS:aggAsymm_cent`), bins = 100)
ggsave(paste0(figDir,"DM_v_mean__LPSagAsymm_pvalue.png"))
ggplot() +
  geom_histogram(aes(x = res_full_aggRecDF$`p_value_trtLPS:aggRec_cent`), bins = 100)
ggsave(paste0(figDir,"DM_v_mean__LPSagRec_pvalue.png"))
ggplot() +
  geom_histogram(aes(x = res_full_groomDF$`p_value_trtLPS:groom_cent`), bins = 100)
ggsave(paste0(figDir,"DM_v_mean__LPSgroom_pvalue.png"))

meanVector <- as.vector(t(sge_mean_matrix_nested))
ggplot() +
  geom_histogram(aes(x = meanVector), bins= 50) + scale_y_log10()

```

compare elo with other behaviors in DM in clusters

```{r cluster 1}
sge_LPS_DM_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c01_DM_matrix')
sge_LPS_mean_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c01_mean_matrix')
sge_LPS_var_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c01_var_matrix')
sge_LPS_pbMean_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c01_pbMean_matrix')
sge_LPS_DMpb_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c01_DMpb_matrix')

sge_NC_DM_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c01_DM_matrix')
sge_NC_mean_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c01_mean_matrix')
sge_NC_var_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c01_var_matrix')
sge_NC_pbMean_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c01_pbMean_matrix')
sge_NC_DMpb_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c01_DMpb_matrix')

## remove genes w/NAas DM values for more than 5 individuals, from NC & LPS DM

numberNAs <- vector()

for (r in 1:dim(sge_NC_DM_matrix)[1]) {
  numberNAs[r] <- sum(is.na(sge_NC_DM_matrix[r,]))
}
sge_NC_DM_matrix <- sge_NC_DM_matrix[numberNAs < 6,]

numberNAs <- vector()
for (r in 1:dim(sge_LPS_DM_matrix)[1]) {
  numberNAs[r] <- sum(is.na(sge_LPS_DM_matrix[r,]))
}
sge_LPS_DM_matrix <- sge_LPS_DM_matrix[numberNAs < 6,]


## only use common set of genes - must exist in all matrices (8)
geneList <- rownames(table(c(rownames(sge_NC_DM_matrix),
            rownames(sge_NC_var_matrix),
            rownames(sge_NC_mean_matrix),
            rownames(sge_NC_pbMean_matrix),
            rownames(sge_NC_DMpb_matrix),
            rownames(sge_LPS_DM_matrix),
            rownames(sge_LPS_var_matrix),
            rownames(sge_LPS_mean_matrix),
            rownames(sge_LPS_pbMean_matrix),
            rownames(sge_LPS_DMpb_matrix))))[table(c(rownames(sge_NC_DM_matrix),
            rownames(sge_NC_var_matrix),
            rownames(sge_NC_mean_matrix),
            rownames(sge_NC_pbMean_matrix),
            rownames(sge_NC_DMpb_matrix),
            rownames(sge_LPS_DM_matrix),
            rownames(sge_LPS_var_matrix),
            rownames(sge_LPS_mean_matrix),
            rownames(sge_LPS_pbMean_matrix),
            rownames(sge_LPS_DMpb_matrix))) == 10]

geneList <- geneList[geneList != "chrMT"]

## subset mean and variance data frames to include only the genes present in the DM data frame ##
sge_NC_DM_matrix <- subset(sge_NC_DM_matrix, rownames(sge_NC_DM_matrix) %in% geneList)
sge_NC_pbMean_matrix <- subset(sge_NC_pbMean_matrix, rownames(sge_NC_pbMean_matrix) %in% geneList)
sge_NC_var_matrix <- subset(sge_NC_var_matrix, rownames(sge_NC_var_matrix) %in% geneList)
sge_NC_mean_matrix <- subset(sge_NC_mean_matrix, rownames(sge_NC_mean_matrix) %in% geneList)
sge_NC_DMpb_matrix <- subset(sge_NC_DMpb_matrix, rownames(sge_NC_DMpb_matrix) %in% geneList)

sge_LPS_DM_matrix <- subset(sge_LPS_DM_matrix, rownames(sge_LPS_DM_matrix) %in% geneList)
sge_LPS_mean_matrix <- subset(sge_LPS_mean_matrix, rownames(sge_LPS_mean_matrix) %in% geneList)
sge_LPS_var_matrix <- subset(sge_LPS_var_matrix, rownames(sge_LPS_var_matrix) %in% geneList)
sge_LPS_pbMean_matrix <- subset(sge_LPS_pbMean_matrix, rownames(sge_LPS_pbMean_matrix) %in% geneList)
sge_LPS_DMpb_matrix <- subset(sge_LPS_DMpb_matrix, rownames(sge_LPS_DMpb_matrix) %in% geneList)

```

```{r}
p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 1') +
  ylab('variance in gene expression - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 2') +
  ylab('variance in gene expression - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 3') +
  ylab('variance in gene expression - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_mean_ge__by_individual_noLim_c01.png"))

p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 1') +
  ylab('variance in gene expression - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 2') +
  ylab('variance in gene expression - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 3') +
  ylab('variance in gene expression - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_mean_ge__by_individual_yLim250_c01.png"))

p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_DM_matrix[,1])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 1') +
  ylab('DM - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_DM_matrix[,2])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 2') +
  ylab('DM - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_DM_matrix[,3])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 3') +
  ylab('DM - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"DM_v_mean_ge__by_individual_c01.png"))
```
v pbMean instead of raw mean
```{r}
p1 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 1') +
  ylab('variance in gene expression - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 2') +
  ylab('variance in gene expression - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 3') +
  ylab('variance in gene expression - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_pbMean_ge__by_individual_noLim_c01.png"))

p1 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 1') +
  ylab('variance in gene expression - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 2') +
  ylab('variance in gene expression - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 3') +
  ylab('variance in gene expression - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_pbMean_ge__by_individual_yLim250_c01.png"))

p1 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,1],sge_NC_DM_matrix[,1])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 1') +
  ylab('DM - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,2],sge_NC_DM_matrix[,2])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 2') +
  ylab('DM - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,3],sge_NC_DM_matrix[,3])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 3') +
  ylab('DM - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"DM_v_pbMean_ge__by_individual_c01.png"))
```


log scale the mean
```{r}
# + scale_x_continuous(trans = "log2")

p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 1') +
  ylab('variance in gene expression - individual 1') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 2') +
  ylab('variance in gene expression - individual 2') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 3') +
  ylab('variance in gene expression - individual 3') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_log2_mean_ge__by_individual_c01.png"))


p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 1') +
  ylab('variance in gene expression - individual 1') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 2') +
  ylab('variance in gene expression - individual 2') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 3') +
  ylab('variance in gene expression - individual 3') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_log2_mean_ge__by_individual_yLim250_c01.png"))

p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_DM_matrix[,1])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 1') +
  ylab('DM - individual 1') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_DM_matrix[,2])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 2') +
  ylab('DM - individual 2') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_DM_matrix[,3])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 3') +
  ylab('DM - individual 3') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"DM_v_log2_mean_ge__by_individual_c01.png"))


```

v pbMean instead of raw mean
```{r}
p1 <- ggplot(sge_NC_DMpb_matrix, aes(sge_NC_DMpb_matrix[,1],sge_NC_DM_matrix[,1])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('DM - individual 1') +
  ylab('DM from pbMean - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_DMpb_matrix, aes(sge_NC_DMpb_matrix[,2],sge_NC_DM_matrix[,2])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('DM - individual 2') +
  ylab('DM from pbMean - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_DMpb_matrix, aes(sge_NC_DMpb_matrix[,3],sge_NC_DM_matrix[,3])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('DM - individual 3') +
  ylab('DM from pbMean - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_pbMean_ge__by_individual_noLim_c01.png"))

```


```{r nest the matrices}

## generate a dataframe of NC and LPS combined ##
sge_DM_matrix_nested <- merge(sge_NC_DM_matrix,sge_LPS_DM_matrix, by=0)
rownames(sge_DM_matrix_nested) <- sge_DM_matrix_nested$Row.names
sge_DM_matrix_nested$Row.names <- NULL

```


```{r fix metadata c01}
## regress out group effects from DM ##
design <- model.matrix(~metadata$group)
fit <-lmFit(sge_DM_matrix_nested,design)
intercepts=data.frame(eBayes(fit))[,1]
resid_DM=apply(residuals.MArrayLM(object=fit, sge_DM_matrix_nested),2,function(x){x+intercepts})

## run a nested model with trt, and age and elo nested within trt ##
## elo and age have to be mean centered for nested model ##
metadata$elo_centered <- scale(metadata$elo, center = T, scale = F)
#metadata$elo_centered <- metadata$elo - mean(metadata$elo)
metadata$age_centered <- scale(metadata$age, center = T, scale = F)
#metadata$age_centered <- metadata$age - mean(metadata$age)

behaveData <- read.table("~/Dropbox (Personal)/scRNA/SGE_DM/data/SGEII_METADATA_MASTER.txt", header = T)

#phase 1 only
behaveData <- subset(behaveData, phase == 1)

#animals we need data for
# behave data is 2 character upper, genomic data is 4 char upper/lower
needIDs <- to_any_case(unique(substr(metadata$ID, start = 1, stop  = 2)), case = "screaming_snake")

#limit behavioral data to correct 45 individs
behaveData <- subset(behaveData, ID %in% needIDs)

#sort by ID
behaveData <- behaveData[order(behaveData$ID),]

#match size of metadata object, repeats for NC & LPS
behaveData <- rbind(behaveData,behaveData)

#get the important variables, centered
metadata$aggAsymm_cent <- behaveData$gc.agg.Diff
metadata$aggRec_cent <- behaveData$gc.overall.agg.rec
metadata$groom_cent <- behaveData$gc.overall.groom


```

Fit an elo model for each gene using emmreml
```{r}
design = model.matrix(~trt+trt:elo_centered+trt:age_centered,data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_DM[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_DM[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}



for(i in 1:nrow(resid_DM))
{
  emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_fullDFc01 <- data.frame(res_full)

plotModelHists(res_fullDFc01, folderName = figDir, modelName = "DM_model_c01")
```


aggAsymm
```{r}
design = model.matrix(~trt+trt:aggAsymm_cent + trt:age_centered, data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_DM[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_DM[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}

for(i in 1:nrow(resid_DM))
{
  emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_full_aggAsymDF_c01 <- as.data.frame(res_full)

plotModelHists(res_full_aggAsymDF_c01, folderName = figDir, modelName = "DM_model_AgAsym_c01")
```

Aggression Received
```{r}
design = model.matrix(~trt+trt:aggRec_cent + trt:age_centered, data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_DM[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_DM[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}

for(i in 1:nrow(resid_DM))
{
  emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_full_aggRecDF_c01 <- as.data.frame(res_full)
plotModelHists(res_full_aggRecDF_c01, folderName = figDir, modelName = "DM_model_AgRec_c01")
```

Overall Grooming
```{r}
design = model.matrix(~trt+trt:groom_cent + trt:age_centered, data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_DM[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_DM[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}

for(i in 1:nrow(resid_DM))
{
  emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_full_groomDF_c01 <- as.data.frame(res_full)
plotModelHists(res_full_groomDF_c01, folderName = figDir, modelName = "DM_model_groom_c01")
```


```{r cluster 2}
sge_LPS_DM_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c02_DM_matrix')
sge_LPS_mean_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c02_mean_matrix')
sge_LPS_var_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c02_var_matrix')

sge_NC_DM_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c02_DM_matrix')
sge_NC_mean_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c02_mean_matrix')
sge_NC_var_matrix <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c02_var_matrix')


## remove genes w/0 as DM values for every individual
## don't have to remove them from 
sge_LPS_DM_matrix <- sge_LPS_DM_matrix[rowSums(sge_LPS_DM_matrix) != 0,]
sge_NC_DM_matrix <- sge_NC_DM_matrix[rowSums(sge_NC_DM_matrix) != 0,]

## remove chrMT counts
#sge_LPS_DM_matrix <- sge_LPS_DM_matrix[rownames(sge_LPS_DM_matrix) != "chrMT",]
#sge_NC_DM_matrix <- sge_NC_DM_matrix[rownames(sge_NC_DM_matrix) != "chrMT",]

## only use common set of genes - must exist in all matrices (8)
geneList <- rownames(table(c(rownames(sge_NC_DM_matrix),
            rownames(sge_NC_var_matrix),
            rownames(sge_NC_mean_matrix),
            rownames(sge_LPS_DM_matrix),
            rownames(sge_LPS_var_matrix),
            rownames(sge_LPS_mean_matrix))))[table(c(rownames(sge_NC_DM_matrix),
            rownames(sge_NC_var_matrix),
            rownames(sge_NC_mean_matrix),
            rownames(sge_LPS_DM_matrix),
            rownames(sge_LPS_var_matrix),
            rownames(sge_LPS_mean_matrix))) == 6]

geneList <- geneList[geneList != "chrMT"]

## subset mean and variance data frames to include only the genes present in the DM data frame ##
sge_NC_DM_matrix <- subset(sge_NC_DM_matrix, rownames(sge_NC_DM_matrix) %in% geneList)
sge_NC_mean_matrix <- subset(sge_NC_mean_matrix, rownames(sge_NC_mean_matrix) %in% geneList)
sge_NC_var_matrix <- subset(sge_NC_var_matrix, rownames(sge_NC_var_matrix) %in% geneList)


sge_LPS_DM_matrix <- subset(sge_LPS_DM_matrix, rownames(sge_LPS_DM_matrix) %in% geneList)
sge_LPS_mean_matrix <- subset(sge_LPS_mean_matrix, rownames(sge_LPS_mean_matrix) %in% geneList)
sge_LPS_var_matrix <- subset(sge_LPS_var_matrix, rownames(sge_LPS_var_matrix) %in% geneList)


p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 1') +
  ylab('variance in gene expression - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 2') +
  ylab('variance in gene expression - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 3') +
  ylab('variance in gene expression - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_mean_ge__by_individual_noLim_c02.png"))

p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 1') +
  ylab('variance in gene expression - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 2') +
  ylab('variance in gene expression - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 3') +
  ylab('variance in gene expression - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_mean_ge__by_individual_yLim250_c02.png"))

p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_DM_matrix[,1])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 1') +
  ylab('DM - individual 1') +
  theme_minimal()
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_DM_matrix[,2])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 2') +
  ylab('DM - individual 2') +
  theme_minimal()
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_DM_matrix[,3])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 3') +
  ylab('DM - individual 3') +
  theme_minimal()

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"DM_v_mean_ge__by_individual_c02.png"))

p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 1') +
  ylab('variance in gene expression - individual 1') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 2') +
  ylab('variance in gene expression - individual 2') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  #ylim(0,1000) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 3') +
  ylab('variance in gene expression - individual 3') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_log2_mean_ge__by_individual_c02.png"))


p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_var_matrix[,1])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 1') +
  ylab('variance in gene expression - individual 1') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_var_matrix[,2])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 2') +
  ylab('variance in gene expression - individual 2') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_var_matrix[,3])) +
  geom_point(alpha=0.5) +
  ylim(0,250) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black', fullrange = T) +
  xlab('mean ge - individual 3') +
  ylab('variance in gene expression - individual 3') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"var_v_log2_mean_ge__by_individual_yLim250_c02.png"))

p1 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,1],sge_NC_DM_matrix[,1])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 1') +
  ylab('DM - individual 1') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p2 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,2],sge_NC_DM_matrix[,2])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 2') +
  ylab('DM - individual 2') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))
p3 <- ggplot(sge_NC_mean_matrix, aes(sge_NC_mean_matrix[,3],sge_NC_DM_matrix[,3])) +
  geom_point(alpha=0.5) +
  #xlim(0,50) +
  stat_smooth(method='lm', color='black') +
  xlab('mean ge - individual 3') +
  ylab('DM - individual 3') +
  theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x)))

plot_grid(p1,p2,p3, ncol = 3)
ggsave(paste0(figDir,"DM_v_log2_mean_ge__by_individual_c02.png"))

## generate a dataframe of NC and LPS combined ##
sge_DM_matrix_nested <- merge(sge_NC_DM_matrix,sge_LPS_DM_matrix, by=0)
rownames(sge_DM_matrix_nested) <- sge_DM_matrix_nested$Row.names
sge_DM_matrix_nested$Row.names <- NULL

## regress out group effects from DM ##
design <- model.matrix(~metadata$group)
fit <-lmFit(sge_DM_matrix_nested,design)
intercepts=data.frame(eBayes(fit))[,1]
resid_DM=apply(residuals.MArrayLM(object=fit, sge_DM_matrix_nested),2,function(x){x+intercepts})

## run a nested model with trt, and age and elo nested within trt ##
## elo and age have to be mean centered for nested model ##
metadata$elo_centered <- scale(metadata$elo, center = T, scale = F)
#metadata$elo_centered <- metadata$elo - mean(metadata$elo)
metadata$age_centered <- scale(metadata$age, center = T, scale = F)
#metadata$age_centered <- metadata$age - mean(metadata$age)

behaveData <- read.table("~/Dropbox (Personal)/scRNA/SGE_DM/data/SGEII_METADATA_MASTER.txt", header = T)

#phase 1 only
behaveData <- subset(behaveData, phase == 1)

#animals we need data for
# behave data is 2 character upper, genomic data is 4 char upper/lower
needIDs <- to_any_case(unique(substr(metadata$ID, start = 1, stop  = 2)), case = "screaming_snake")

#limit behavioral data to correct 45 individs
behaveData <- subset(behaveData, ID %in% needIDs)

#sort by ID
behaveData <- behaveData[order(behaveData$ID),]

#match size of metadata object, repeats for NC & LPS
behaveData <- rbind(behaveData,behaveData)

#get the important variables, centered
metadata$aggAsymm_cent <- behaveData$gc.agg.Diff
metadata$aggRec_cent <- behaveData$gc.overall.agg.rec
metadata$groom_cent <- behaveData$gc.overall.groom


## elo model

design = model.matrix(~trt+trt:elo_centered+trt:age_centered,data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_DM[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_DM[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}



for(i in 1:nrow(resid_DM))
{
  emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_fullDFc02 <- data.frame(res_full)

plotModelHists(res_fullDFc02, folderName = figDir, modelName = "DM_model_c02")



##aggAsymm

design = model.matrix(~trt+trt:aggAsymm_cent + trt:age_centered, data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_DM[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_DM[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}

for(i in 1:nrow(resid_DM))
{
  emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_full_aggAsymDF_c02 <- as.data.frame(res_full)

plotModelHists(res_full_aggAsymDF_c02, folderName = figDir, modelName = "DM_model_AgAsym_c02")


##Aggression Received

design = model.matrix(~trt+trt:aggRec_cent + trt:age_centered, data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_DM[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_DM[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}

for(i in 1:nrow(resid_DM))
{
  emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_full_aggRecDF_c02 <- as.data.frame(res_full)
plotModelHists(res_full_aggRecDF_c02, folderName = figDir, modelName = "DM_model_AgRec_c02")

##grooming

design = model.matrix(~trt+trt:groom_cent + trt:age_centered, data=metadata)

#Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
res_full=resid_DM[,1:(3*ncol(design))]
colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))

#Declare object random_effects to store in it the individual-wise u random effects
random_effects=resid_DM[,1:45]
colnames(random_effects)=metadata[1:45,'ID']

#Define matrix Z describing the sample-to-individual mapping
K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
# correcting capitalization so it matches
rownames(K)[34] <- 'JE11'
colnames(K)[34] <- 'JE11'
K=K[unique(metadata$ID),unique(metadata$ID)]

Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
rownames(Z)=metadata$ID
colnames(Z)=colnames(K)
for(i in 1:ncol(Z)){
  set=which(metadata$ID == colnames(Z)[i])
  Z[set,i]=1
}

for(i in 1:nrow(resid_DM))
{
  emma=emmreml(y=resid_DM[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
  random_effects[i,]=t(emma$uhat)
  res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
}

res_full_groomDF_c02 <- as.data.frame(res_full)
plotModelHists(res_full_groomDF_c02, folderName = figDir, modelName = "DM_model_groom_c02")
```