---
title: "rankEffect_DM_modeling_clusters"
author: "c ryan"
date: "1/29/2021"
output: html_document
---

load helpful functions
```{r}
perm.fdr=function(input_df,perm_df,Pvals_col_name,name){
  
  pvals_index=which(colnames(input_df)==Pvals_col_name)
  ro<-input_df[order(input_df[,pvals_index]),]
  p_obs <- data.frame(pvalue=ro[,pvals_index])
  p_vector<-matrix(as.matrix(perm_df),ncol=1)
  p_vector=data.frame(p_vector[order(p_vector)])
  
  F<-p_obs[,1]
  F_o<-p_obs[,1]
  pi_hat<-p_obs[,1]
  
  j=1
  observed<-length(p_obs[,1])
  randoms<-length(p_vector[,1])
  
  for(i in 1:observed)
  {
    repeat
    {
      if((p_vector[j,1]<p_obs[i,1])&j<randoms){j<-j+1}else{break}
    }
    F[i]=i/observed
    F_o[i]=(j-1)/randoms
    if(F_o[i]<1){pi_hat[i]=(1-F[i])/(1-F_o[i])}else{pi_hat[i]=1}
  }
  tabla <-data.frame(pi_hat,pval=p_obs[,1])
  
  tabla[1,]=c(1,0)
  last_percentile_average=mean(tabla$pi_hat[as.integer(min((length(tabla[,1])*0.99),(nrow(tabla)-1)):length(tabla[,1]))])
  tabla[nrow(tabla),]=c(last_percentile_average,1)
  constraint_matrix=as.matrix(data.frame(c(0,2),c(0,1),c(1,0)))
  f_hat<-suppressWarnings(cobs(tabla$pval,tabla$pi_hat,constraint="convex",pointwise=constraint_matrix,maxiter=1000,print.warn=FALSE,print.mesg=FALSE))
  
  f_hat_serie=f_hat$fitted
  pi_o=f_hat_serie[length(f_hat_serie)]
  pi_o=min(pi_o,1)
  pi_o=max(pi_o,0)
  
  Fdr_ST_perm=pi_o*F_o/F
  
  for(i in 1:length(p_obs[,1]))
  {
    Fdr_ST_perm[i]=pi_o*F_o[i]/F[i]
    if(i>1)
    {
      for(j in 1:(i-1))
      {
        if(Fdr_ST_perm[i-j]>Fdr_ST_perm[i]){Fdr_ST_perm[i-j]=Fdr_ST_perm[i]}else{break}
      }
    }
    if(Fdr_ST_perm[i]>1)  Fdr_ST_perm[i]=1
  }
  
  fdrs_df <-data.frame(ro,q_ST_perm=Fdr_ST_perm)
  rownames(fdrs_df)=rownames(ro)
  colnames(fdrs_df)[ncol(fdrs_df)]=paste0("fdr_",name)
  
  return(fdrs_df)
}


plotModelHists <- function(emmremlDF, folderName = "~/Downloads/", modelName = "model") {
  #modelName <- "DMpb_model_AgAsym_c02"
  #emmremlDF <- res_full_aggAsym_DMpb_c02
  #folderName <- figDir
  
  nVars <- dim(emmremlDF)[2] / 3
  
  colNums <- (2 * nVars + 1) : (3 * nVars)
  
  for (c in colNums) {
    ggplot() +
      geom_histogram(aes(x = emmremlDF[,c]), bins = 100) +
      xlab(colnames(emmremlDF)[c]) + ggtitle(paste0("emmreml run ",modelName))
    ggsave(paste0(folderName,"emmreml_hist_",modelName,"_",str_replace(colnames(emmremlDF)[c],":","_"),".png"), width = 6, height = 4)
  }
}

```

```{r packages, include=FALSE}
library(ggplot2)
library(limma)
library(edgeR)
library(EMMREML)
library(cobs)
library(cowplot)
library(snakecase)
library(stringr)
library(qvalue)
library(scales)
library(reshape2)
library(topGO)
library(biomaRt)

figDir <- "~/Dropbox (Personal)/scRNA/SGE_DM/figures/"
workDir <- "~/Dropbox (Personal)/scRNA/SGE_DM/"
```


```{r loop through clusters to test variance and DM}

for (cluster in c("00","14","02","03","04","05","06","07")) {
  cluster <- "01"
  
  sge_LPS_DM_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c',cluster,'_DM_matrix'))
  sge_LPS_mean_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c',cluster,'_mean_matrix'))
  sge_LPS_var_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c',cluster,'_var_matrix'))
  sge_LPS_pbMean_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c',cluster,'_pbMean_matrix'))
  sge_LPS_DMpb_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c',cluster,'_DMpb_matrix'))
  
  sge_NC_DM_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c',cluster,'_DM_matrix'))
  sge_NC_mean_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c',cluster,'_mean_matrix'))
  sge_NC_var_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c',cluster,'_var_matrix'))
  sge_NC_pbMean_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c',cluster,'_pbMean_matrix'))
  sge_NC_DMpb_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c',cluster,'_DMpb_matrix'))
  
  ## remove genes w/NAas DM values for any individuals, from NC & LPS DM
  numberNAs <- vector()
  
  for (r in 1:dim(sge_NC_DM_matrix)[1]) {
    numberNAs[r] <- sum(is.na(sge_NC_DM_matrix[r,]))
  }
  sge_NC_DM_matrix <- sge_NC_DM_matrix[numberNAs < 1,]
  
  numberNAs <- vector()
  for (r in 1:dim(sge_LPS_DM_matrix)[1]) {
    numberNAs[r] <- sum(is.na(sge_LPS_DM_matrix[r,]))
  }
  sge_LPS_DM_matrix <- sge_LPS_DM_matrix[numberNAs < 1,]
  
  
  ## only use common set of genes - must exist in all matrices (8)
  geneList <- rownames(table(c(rownames(sge_NC_DM_matrix),
              rownames(sge_NC_var_matrix),
              rownames(sge_NC_mean_matrix),
              rownames(sge_NC_pbMean_matrix),
              rownames(sge_NC_DMpb_matrix),
              rownames(sge_LPS_DM_matrix),
              rownames(sge_LPS_var_matrix),
              rownames(sge_LPS_mean_matrix),
              rownames(sge_LPS_pbMean_matrix),
              rownames(sge_LPS_DMpb_matrix))))[table(c(rownames(sge_NC_DM_matrix),
              rownames(sge_NC_var_matrix),
              rownames(sge_NC_mean_matrix),
              rownames(sge_NC_pbMean_matrix),
              rownames(sge_NC_DMpb_matrix),
              rownames(sge_LPS_DM_matrix),
              rownames(sge_LPS_var_matrix),
              rownames(sge_LPS_mean_matrix),
              rownames(sge_LPS_pbMean_matrix),
              rownames(sge_LPS_DMpb_matrix))) == 10]
  
  geneList <- geneList[geneList != "chrMT"]
  
  
  # sge_NC_DM_matrix <- subset(sge_NC_DM_matrix, rownames(sge_NC_DM_matrix) %in% geneList)
  # sge_NC_pbMean_matrix <- subset(sge_NC_pbMean_matrix, rownames(sge_NC_pbMean_matrix) %in% geneList)
  # sge_NC_var_matrix <- subset(sge_NC_var_matrix, rownames(sge_NC_var_matrix) %in% geneList)
  # sge_NC_mean_matrix <- subset(sge_NC_mean_matrix, rownames(sge_NC_mean_matrix) %in% geneList)
  # sge_NC_DMpb_matrix <- subset(sge_NC_DMpb_matrix, rownames(sge_NC_DMpb_matrix) %in% geneList)
  # 
  # sge_LPS_DM_matrix <- subset(sge_LPS_DM_matrix, rownames(sge_LPS_DM_matrix) %in% geneList)
  # sge_LPS_mean_matrix <- subset(sge_LPS_mean_matrix, rownames(sge_LPS_mean_matrix) %in% geneList)
  # sge_LPS_var_matrix <- subset(sge_LPS_var_matrix, rownames(sge_LPS_var_matrix) %in% geneList)
  # sge_LPS_pbMean_matrix <- subset(sge_LPS_pbMean_matrix, rownames(sge_LPS_pbMean_matrix) %in% geneList)
  # sge_LPS_DMpb_matrix <- subset(sge_LPS_DMpb_matrix, rownames(sge_LPS_DMpb_matrix) %in% geneList)
  
  
  #for any objects that start sge_NC or sge_LP, cut down their gene list and replace 77_06 with Mo
  
  for (fObj in apropos("^sge_[NL][CP]")) {
    tmpObj <- get(fObj)
    ## subset mean and variance data frames to include only the genes present in the DM data frame ##
    tmpObj <- subset(tmpObj, rownames(tmpObj) %in% geneList)
    
    ## fix MO - 77_06 ID
    if (sum(colnames(tmpObj) == "X77_06") > 0) {
      colnames(tmpObj)[which(colnames(tmpObj) == "X77_06")] <- "Mo"
    }
    #sort the columns by ID to match metadata
    tmpObj <- tmpObj[,order(colnames(tmpObj))]
    
    #reassign the temp object to the sge_NC, etc
    assign(x = fObj, value = tmpObj)
  }
  
  
  ## generate a dataframe of NC and LPS combined ##
  sge_DM_matrix_nested <- merge(sge_NC_DM_matrix,sge_LPS_DM_matrix, by=0)
  rownames(sge_DM_matrix_nested) <- sge_DM_matrix_nested$Row.names
  sge_DM_matrix_nested$Row.names <- NULL
  
  #one for pbMeans as well
  sge_pbMean_matrix_nested <- merge(sge_NC_pbMean_matrix,sge_LPS_pbMean_matrix, by=0)
  rownames(sge_pbMean_matrix_nested) <- sge_pbMean_matrix_nested$Row.names
  sge_pbMean_matrix_nested$Row.names <- NULL
  
  #and raw means and variances to use in the variance model
  sge_mean_matrix_nested <- merge(sge_NC_mean_matrix,sge_LPS_mean_matrix, by=0)
  rownames(sge_mean_matrix_nested) <- sge_mean_matrix_nested$Row.names
  sge_mean_matrix_nested$Row.names <- NULL
  
  sge_var_matrix_nested <- merge(sge_NC_var_matrix,sge_LPS_var_matrix, by=0)
  rownames(sge_var_matrix_nested) <- sge_var_matrix_nested$Row.names
  sge_var_matrix_nested$Row.names <- NULL
  
  sge_DMpb_matrix_nested <- merge(sge_NC_DMpb_matrix,sge_LPS_DMpb_matrix, by=0)
  rownames(sge_DMpb_matrix_nested) <- sge_DMpb_matrix_nested$Row.names
  sge_DMpb_matrix_nested$Row.names <- NULL
  
  
  indIDs <- colnames(sge_NC_DM_matrix)

  #pick metadata based on samples in the matrix
  
  ## Here I'm doing some modeling to ask what are the effects of treatment and rank on DM ##
  ## get some metadata ##
  metadata <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_metadata_matrix', header = T) # this files is also in SGE_DM/dataFiles
  metadata[1,1] <- 'Mo'
  metadata <- metadata[order(metadata$ID),]
  
  metadata <- metadata[metadata$ID %in% indIDs,]
  
  ## make sure metadata order matches data frames ##
  all.equal(metadata$ID,colnames(sge_NC_DM_matrix))
  
  ## double the metadata so it includes both NC and LPS ##
  metadata <- rbind(metadata,metadata)
  metadata$trt <- rep(c('NC','LPS'), times=c(length(indIDs),length(indIDs)))
  
  ## run a nested model with trt, and age and elo nested within trt ##
  ## elo and age have to be mean centered for nested model ##
  metadata$elo_centered <- scale(metadata$elo, center = T, scale = F)
  #metadata$elo_centered <- metadata$elo - mean(metadata$elo)
  metadata$age_centered <- scale(metadata$age, center = T, scale = F)
  #metadata$age_centered <- metadata$age - mean(metadata$age)
  
  behaveData <- read.table("~/Dropbox (Personal)/scRNA/SGE_DM/data/SGEII_METADATA_MASTER.txt", header = T)
  
  #phase 1 only
  behaveData <- subset(behaveData, phase == 1)
  
  #animals we need data for
  # behave data is 2 character upper, genomic data is 4 char upper/lower
  needIDs <- to_any_case(unique(substr(metadata$ID, start = 1, stop  = 2)), case = "screaming_snake")
  
  #limit behavioral data to correct 45 individs
  behaveData <- subset(behaveData, ID %in% needIDs)
  
  #sort by ID
  behaveData <- behaveData[order(behaveData$ID),]
  
  #match size of metadata object, repeats for NC & LPS
  behaveData <- rbind(behaveData,behaveData)
  
  #get the important variables, centered
  metadata$aggAsymm_cent <- behaveData$gc.agg.Diff
  metadata$aggRec_cent <- behaveData$gc.overall.agg.rec
  metadata$groom_cent <- behaveData$gc.overall.groom
  
  ## regress out group effects from DM ##
  design <- model.matrix(~metadata$group)
  fit <-lmFit(sge_DM_matrix_nested,design)
  intercepts=data.frame(eBayes(fit))[,1]
  resid_DM=apply(residuals.MArrayLM(object=fit, sge_DM_matrix_nested),2,function(x){x+intercepts})
  
  ## regress out group effects from DMpb ##
  design <- model.matrix(~metadata$group)
  fit <-lmFit(sge_DMpb_matrix_nested,design)
  intercepts=data.frame(eBayes(fit))[,1]
  resid_DMpb=apply(residuals.MArrayLM(object=fit, sge_DMpb_matrix_nested),2,function(x){x+intercepts})
  
  ## regress out group effects from pbMean ##
  design <- model.matrix(~metadata$group)
  fit <-lmFit(sge_pbMean_matrix_nested,design)
  intercepts=data.frame(eBayes(fit))[,1]
  resid_pbMean=apply(residuals.MArrayLM(object=fit, sge_pbMean_matrix_nested),2,function(x){x+intercepts})
  
  
  #Variance v pbMean instead of raw mean
  
  p1 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,1],sge_NC_var_matrix[,1])) +
    geom_point(alpha=0.5) +
    ylim(0,250) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    xlab('mean ge - individual 1') +
    ylab('variance in gene expression - individual 1') +
    theme_minimal()
  p2 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,2],sge_NC_var_matrix[,2])) +
    geom_point(alpha=0.5) +
    ylim(0,250) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    xlab('mean ge - individual 2') +
    ylab('variance in gene expression - individual 2') +
    theme_minimal()
  p3 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,3],sge_NC_var_matrix[,3])) +
    geom_point(alpha=0.5) +
    ylim(0,250) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    xlab('mean ge - individual 3') +
    ylab('variance in gene expression - individual 3') +
    theme_minimal()
  
  plot_grid(p1,p2,p3, ncol = 3)
  ggsave(paste0(figDir,"c",cluster,"/","var_v_pbMean_ge__by_individual_yLim250_c",cluster,".png"))
  
  #DM v pbMean
  
  p1 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,1],sge_NC_DM_matrix[,1])) +
    geom_point(alpha=0.5) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black') +
    xlab('mean ge - individual 1') +
    ylab('DM - individual 1') +
    theme_minimal()
  p2 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,2],sge_NC_DM_matrix[,2])) +
    geom_point(alpha=0.5) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black') +
    xlab('mean ge - individual 2') +
    ylab('DM - individual 2') +
    theme_minimal()
  p3 <- ggplot(sge_NC_pbMean_matrix, aes(sge_NC_pbMean_matrix[,3],sge_NC_DM_matrix[,3])) +
    geom_point(alpha=0.5) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black') +
    xlab('mean ge - individual 3') +
    ylab('DM - individual 3') +
    theme_minimal()
  
  plot_grid(p1,p2,p3, ncol = 3)
  ggsave(paste0(figDir,"c",cluster,"/","DM_v_pbMean_ge__by_individual_c",cluster,".png"))
  
  #v pbMean instead of raw mean
  
  p1 <- ggplot(sge_NC_DMpb_matrix, aes(sge_NC_DMpb_matrix[,1],sge_NC_DM_matrix[,1])) +
    geom_point(alpha=0.5) +
    #ylim(0,1000) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    ylab('DM - individual 1') +
    xlab('DM from pbMean - individual 1') +
    theme_minimal()
  p2 <- ggplot(sge_NC_DMpb_matrix, aes(sge_NC_DMpb_matrix[,2],sge_NC_DM_matrix[,2])) +
    geom_point(alpha=0.5) +
    #ylim(0,1000) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    ylab('DM - individual 2') +
    xlab('DM from pbMean - individual 2') +
    theme_minimal()
  p3 <- ggplot(sge_NC_DMpb_matrix, aes(sge_NC_DMpb_matrix[,3],sge_NC_DM_matrix[,3])) +
    geom_point(alpha=0.5) +
    #ylim(0,1000) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    ylab('DM - individual 3') +
    xlab('DM from pbMean - individual 3') +
    theme_minimal()
  
  plot_grid(p1,p2,p3, ncol = 3)
  ggsave(paste0(figDir,"c",cluster,"/","DMpb_v_DM_ge__by_individual_noLim_c",cluster,".png"))
  
  
  #var v DM
  p1 <- ggplot(sge_NC_var_matrix, aes(sge_NC_var_matrix[,1],sge_NC_DM_matrix[,1])) +
    geom_point(alpha=0.5) +
    #ylim(0,1000) +
    xlim(0,250) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    ylab('DM - individual 1') +
    xlab('var - individual 1') +
    theme_minimal()
  p2 <- ggplot(sge_NC_var_matrix, aes(sge_NC_var_matrix[,2],sge_NC_DM_matrix[,2])) +
    geom_point(alpha=0.5) +
    #ylim(0,1000) +
    xlim(0,250) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    ylab('DM - individual 2') +
    xlab('var - individual 2') +
    theme_minimal()
  p3 <- ggplot(sge_NC_var_matrix, aes(sge_NC_var_matrix[,3],sge_NC_DM_matrix[,3])) +
    geom_point(alpha=0.5) +
    #ylim(0,1000) +
    xlim(0,250) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    ylab('DM - individual 3') +
    xlab('var - individual 3') +
    theme_minimal()
  
  plot_grid(p1,p2,p3, ncol = 3)
  ggsave(paste0(figDir,"c",cluster,"/","var_v_DM_ge__by_individual_xLim250_c",cluster,".png"))
  
  ###### Fit an elo model for each gene using emmreml
  
  ### modeling DMpb
  
  ###### Fit an elo model for each gene using emmreml to DM pseudobulk
  
  design = model.matrix(~trt+trt:elo_centered+trt:age_centered,data=metadata)
  
  #Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
  res_full=resid_DMpb[,1:(3*ncol(design))]
  colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
  colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
  colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))
  
  #Declare object random_effects to store in it the individual-wise u random effects
  random_effects=resid_DMpb[,1:length(indIDs)]
  colnames(random_effects)=metadata[1:length(indIDs),'ID']
  
  #Define matrix Z describing the sample-to-individual mapping
  K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
  # correcting capitalization so it matches
  rownames(K)[34] <- 'JE11'
  colnames(K)[34] <- 'JE11'
  K=K[unique(metadata$ID),unique(metadata$ID)]
  
  Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
  rownames(Z)=metadata$ID
  colnames(Z)=colnames(K)
  for(i in 1:ncol(Z)){
    set=which(metadata$ID == colnames(Z)[i])
    Z[set,i]=1
  }
  
  for(i in 1:nrow(resid_DMpb))
  {
    emma=emmreml(y=resid_DMpb[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
    random_effects[i,]=t(emma$uhat)
    res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
  }
  
  assign(value = data.frame(res_full), x = paste0("res_full_DMpb_c",cluster))
  
  plotModelHists(get(paste0("res_full_DMpb_c",cluster)), folderName = paste0(figDir,"c",cluster,"/"), modelName = paste0("DMpb_model_c",cluster))
  
  ######Now the other behavioral metrics
  ######agAsymm | AgRec | Groom
  
  #Agonism Asymmetry
  design = model.matrix(~trt+trt:aggAsymm_cent + trt:age_centered, data=metadata)
  
  #Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
  res_full=resid_DMpb[,1:(3*ncol(design))]
  colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
  colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
  colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))
  
  #Declare object random_effects to store in it the individual-wise u random effects
  random_effects=resid_DMpb[,1:length(indIDs)]
  colnames(random_effects)=metadata[1:length(indIDs),'ID']
  
  #Define matrix Z describing the sample-to-individual mapping
  K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
  # correcting capitalization so it matches
  rownames(K)[34] <- 'JE11'
  colnames(K)[34] <- 'JE11'
  K=K[unique(metadata$ID),unique(metadata$ID)]
  
  Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
  rownames(Z)=metadata$ID
  colnames(Z)=colnames(K)
  for(i in 1:ncol(Z)){
    set=which(metadata$ID == colnames(Z)[i])
    Z[set,i]=1
  }
  
  for(i in 1:nrow(resid_DMpb))
  {
    emma=emmreml(y=resid_DMpb[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
    random_effects[i,]=t(emma$uhat)
    res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
  }
  
  assign(value = data.frame(res_full), x = paste0("res_full_aggAsym_DMpb_c",cluster))
  plotModelHists(get(paste0("res_full_aggAsym_DMpb_c",cluster)), folderName = paste0(figDir,"c",cluster,"/"), modelName = paste0("DMpb_model_AgAsym_c",cluster))
  
  
  #Agonism Received
  design = model.matrix(~trt+trt:aggRec_cent + trt:age_centered, data=metadata)
  
  #Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
  res_full=resid_DMpb[,1:(3*ncol(design))]
  colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
  colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
  colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))
  
  #Declare object random_effects to store in it the individual-wise u random effects
  random_effects=resid_DMpb[,1:length(indIDs)]
  colnames(random_effects)=metadata[1:length(indIDs),'ID']
  
  #Define matrix Z describing the sample-to-individual mapping
  K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
  # correcting capitalization so it matches
  rownames(K)[34] <- 'JE11'
  colnames(K)[34] <- 'JE11'
  K=K[unique(metadata$ID),unique(metadata$ID)]
  
  Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
  rownames(Z)=metadata$ID
  colnames(Z)=colnames(K)
  for(i in 1:ncol(Z)){
    set=which(metadata$ID == colnames(Z)[i])
    Z[set,i]=1
  }
  
  for(i in 1:nrow(resid_DMpb))
  {
    emma=emmreml(y=resid_DMpb[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
    random_effects[i,]=t(emma$uhat)
    res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
  }
  
  assign(value = data.frame(res_full), x = paste0("res_full_aggRec_DMpb_c",cluster))
  plotModelHists(get(paste0("res_full_aggRec_DMpb_c",cluster)), folderName = paste0(figDir,"c",cluster,"/"), modelName = paste0("DMpb_model_AgRec_c",cluster))
  
  #Overall Grooming
  
  design = model.matrix(~trt+trt:groom_cent + trt:age_centered, data=metadata)
  
  #Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
  res_full=resid_DMpb[,1:(3*ncol(design))]
  colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
  colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
  colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))
  
  #Declare object random_effects to store in it the individual-wise u random effects
  random_effects=resid_DMpb[,1:length(indIDs)]
  colnames(random_effects)=metadata[1:length(indIDs),'ID']
  
  #Define matrix Z describing the sample-to-individual mapping
  K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
  # correcting capitalization so it matches
  rownames(K)[34] <- 'JE11'
  colnames(K)[34] <- 'JE11'
  K=K[unique(metadata$ID),unique(metadata$ID)]
  
  Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
  rownames(Z)=metadata$ID
  colnames(Z)=colnames(K)
  for(i in 1:ncol(Z)){
    set=which(metadata$ID == colnames(Z)[i])
    Z[set,i]=1
  }
  
  for(i in 1:nrow(resid_DMpb))
  {
    emma=emmreml(y=resid_DMpb[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
    random_effects[i,]=t(emma$uhat)
    res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
  }
  
  assign(value = data.frame(res_full), x = paste0("res_full_groom_DMpb_c",cluster))
  plotModelHists(get(paste0("res_full_groom_DMpb_c",cluster)), folderName = paste0(figDir,"c",cluster,"/"), modelName = paste0("DMpb_model_groom_c",cluster))
  

  ######data validation possibilities
  ######are genes that are upregulated in LPS also higher variance in LPS?
  ###### obtain directly from Paul's data instead
}

```




compare betas from the various models
```{r}
# 
# res_fullDF_pbMean_c01plot <- res_fullDF_pbMean_c01
# 
# pbMeanSig <- qvalue(p = res_fullDF_pbMean_c01$p_value_trtNC, fdr.level = .0000001)$significant
# sum(pbMeanSig)/length(pbMeanSig)
# 
# ggplot() +
#   geom_point(aes(x = res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5,
#                  y = res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5,
#                  col = pbMeanSig),
#              alpha = .1) +
#   xlab("variance model - std betas LPS") +
#   ylab("DM model - std betas LPS") +
#   geom_smooth(aes(x = res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5,
#                  y = res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5), method = "lm") +
#   geom_text(aes(x = .95 * min(res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5),
#                 y = .95 * max(res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5)),
#                 label = paste0("R = ",round(trtR, digits = 3)))
# ggsave(paste0(figDir,"c",cluster,"/","stdBeta_comp_Trt_DM_v_Var_pbMeanSigColor_c",cluster,".png"))
# 
# ggplot() +
#   geom_point(aes(x = res_fullvarDF$beta_trtLPS.elo_centered / (res_fullvarDF$sdev_trtLPS.elo_centered)^0.5,
#                  y = res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5,
#                  col = pbMeanSig),
#              alpha = .1) +
#   xlab("variance model - std betas LPS:elo") +
#   ylab("DM model - std betas LPS:elo") +
#   geom_smooth(aes(x = res_fullvarDF$beta_trtLPS.elo_centered / (res_fullvarDF$sdev_trtLPS.elo_centered)^0.5,
#                  y = res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5), method = "lm") +
#   geom_text(aes(x = .95 * min(res_fullvarDF$beta_trtLPS.elo_centered / (res_fullvarDF$sdev_trtLPS.elo_centered)^0.5),
#                 y = .95 * max(res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5)),
#                 label = paste0("R = ",round(lpsEloR, digits = 3)))
# ggsave(paste0(figDir,"c",cluster,"/","stdBeta_comp_LPSelo_DM_v_Var_pbMeanSigColor_c",cluster,".png"))
# 
# #but what about direction
# res_fullDF_pbMean_c01plot$geneCat <- ifelse(pbMeanSig, ifelse(res_fullDF_pbMean_c01plot$beta_trtNC > 0, "sigDown", "sigUp"),"ns")
# 
# ggplot() +
#   geom_point(aes(x = res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5,
#                  y = res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5,
#                  col = res_fullDF_pbMean_c01plot$geneCat),
#              alpha = .5) +
#   xlab("variance model - std betas LPS") +
#   ylab("DM model - std betas LPS") +
#   geom_smooth(aes(x = res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5,
#                  y = res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5), method = "lm") +
#   geom_text(aes(x = .95 * min(res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5),
#                 y = .95 * max(res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5)),
#                 label = paste0("R = ",round(trtR, digits = 3))) +
#   labs(col = "pbMeans\nSig and\nLPS dir")
# ggsave(paste0(figDir,"c",cluster,"/","stdBeta_comp_Trt_DM_v_Var_pbMeanUpDownColor_c",cluster,".png"))
# 
# ggplot() +
#   geom_point(aes(x = res_fullvarDF$beta_trtLPS.elo_centered / (res_fullvarDF$sdev_trtLPS.elo_centered)^0.5,
#                  y = res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5,
#                  col = res_fullDF_pbMean_c01plot$geneCat),
#              alpha = .1) +
#   xlab("variance model - std betas LPS:elo") +
#   ylab("DM model - std betas LPS:elo") +
#   geom_smooth(aes(x = res_fullvarDF$beta_trtLPS.elo_centered / (res_fullvarDF$sdev_trtLPS.elo_centered)^0.5,
#                  y = res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5), method = "lm") +
#   geom_text(aes(x = .95 * min(res_fullvarDF$beta_trtLPS.elo_centered / (res_fullvarDF$sdev_trtLPS.elo_centered)^0.5),
#                 y = .95 * max(res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5)),
#                 label = paste0("R = ",round(lpsEloR, digits = 3))) +
#   labs(col = "pbMeans\nSig and\nLPS dir")
# ggsave(paste0(figDir,"c",cluster,"/","stdBeta_comp_LPSelo_DM_v_Var_pbMeanUpDownColor_c",cluster,".png"))
# 
# 
# 
# #what if i only plot the significant genes in DM
# res_fullvarDFsig <- res_fullvarDF[qvalue(p = res_fullDF$p_value_trtNC, fdr.level = .1)$significant,]
# res_fullDFsig <- res_fullDF[qvalue(p = res_fullDF$p_value_trtNC, fdr.level = .1)$significant,]
# res_fullDF_pbMean_c01plotSIG <- res_fullDF_pbMean_c01plot[qvalue(p = res_fullDF$p_value_trtNC, fdr.level = .1)$significant,]
# 
# ggplot() +
#   geom_point(aes(x = res_fullvarDFsig$beta_trtNC / (res_fullvarDFsig$sdev_trtNC)^0.5,
#                  y = res_fullDFsig$beta_trtNC / (res_fullDFsig$sdev_trtNC)^0.5,
#                  col = res_fullDF_pbMean_c01plotSIG$geneCat),
#              alpha = .5) +
#   xlab("variance model - std betas LPS") +
#   ylab("DM model - std betas LPS") +
#   labs(col = "pbMeans\nSig and\nLPS dir")
# ggsave(paste0(figDir,"c",cluster,"/","stdBeta_comp_Trt_DM_v_Var_pbMeanUpDownColor_sigDM_c",cluster,".png"))
# 
# 
# #what if i only plot the significant genes in Var
# res_fullvarDFsig <- res_fullvarDF[qvalue(p = res_fullvarDF$p_value_trtNC, fdr.level = .1)$significant,]
# res_fullDFsig <- res_fullDF[qvalue(p = res_fullvarDF$p_value_trtNC, fdr.level = .1)$significant,]
# res_fullDF_pbMean_c01plotSIG <- res_fullDF_pbMean_c01plot[qvalue(p = res_fullvarDF$p_value_trtNC, fdr.level = .1)$significant,]
# 
# ggplot() +
#   geom_point(aes(x = res_fullvarDFsig$beta_trtNC / (res_fullvarDFsig$sdev_trtNC)^0.5,
#                  y = res_fullDFsig$beta_trtNC / (res_fullDFsig$sdev_trtNC)^0.5,
#                  col = res_fullDF_pbMean_c01plotSIG$geneCat),
#              alpha = .5) +
#   xlab("variance model - std betas LPS") +
#   ylab("DM model - std betas LPS") +
#   labs(col = "pbMeans\nSig and\nLPS dir")
# ggsave(paste0(figDir,"c",cluster,"/","stdBeta_comp_Trt_DM_v_Var_pbMeanUpDownColor_sigVar_c",cluster,".png"))
# 
# 
# 
# lpsMeanVarR <- cor.test(res_fullDF_pbMean_c01$beta_trtNC / (res_fullDF_pbMean_c01$sdev_trtNC)^0.5,
#                     res_fullDF_var_c01$beta_trtNC / (res_fullDF_var_c01$sdev_trtNC)^0.5)$estimate
# 
# 
# ggplot() +
#   geom_point(aes(x = res_fullDF_pbMean_c01$beta_trtNC / (res_fullDF_pbMean_c01$sdev_trtNC)^0.5,
#                  y = res_fullDF_var_c01$beta_trtNC / (res_fullDF_var_c01$sdev_trtNC)^0.5),
#              alpha = .1) +
#   xlab("pbMean LPS std beta") +
#   ylab("var LPS std beta") +
#   geom_smooth(aes(x = res_fullDF_pbMean_c01$beta_trtNC / (res_fullDF_pbMean_c01$sdev_trtNC)^0.5,
#                  y = res_fullDF_var_c01$beta_trtNC / (res_fullDF_var_c01$sdev_trtNC)^0.5), method = "lm") +
#   geom_text(aes(x = .95 * min(res_fullDF_pbMean_c01$beta_trtNC / (res_fullDF_pbMean_c01$sdev_trtNC)^0.5),
#                 y = .95 * max(res_fullDF_var_c01$beta_trtNC / (res_fullDF_var_c01$sdev_trtNC)^0.5)),
#                 label = paste0("R = ",round(lpsMeanVarR, digits = 3)))
# 
# ggsave(paste0(figDir,"c",cluster,"/","stdBeta_comp_Trt-var_v_Trt-pbMean_c",cluster,".png"))
# 
# lpsMeanDMR <- cor.test(res_fullDF_pbMean_c01$beta_trtNC / (res_fullDF_pbMean_c01$sdev_trtNC)^0.5,
#                     res_fullDFc01$beta_trtNC / (res_fullDFc01$sdev_trtNC)^0.5)$estimate
# 
# ggplot() +
#   geom_point(aes(x = res_fullDF_pbMean_c01$beta_trtNC / (res_fullDF_pbMean_c01$sdev_trtNC)^0.5,
#                  y = res_fullDFc01$beta_trtNC / (res_fullDFc01$sdev_trtNC)^0.5),
#              alpha = .1) +
#   xlab("pbMean LPS std beta") +
#   ylab("DM LPS std beta") +
#   geom_smooth(aes(x = res_fullDF_pbMean_c01$beta_trtNC / (res_fullDF_pbMean_c01$sdev_trtNC)^0.5,
#                  y = res_fullDFc01$beta_trtNC / (res_fullDFc01$sdev_trtNC)^0.5), method = "lm") +
#   geom_text(aes(x = .95 * min(res_fullDF_pbMean_c01$beta_trtNC / (res_fullDF_pbMean_c01$sdev_trtNC)^0.5),
#                 y = .95 * max(res_fullDFc01$beta_trtNC / (res_fullDFc01$sdev_trtNC)^0.5)),
#                 label = paste0("R = ",round(lpsMeanDMR, digits = 3)))
# ggsave(paste0(figDir,"c",cluster,"/","stdBeta_comp_Trt-DM_v_Trt-pbMean_c",cluster,".png"))

```

how does cluster size affect gene counts
```{r cluster size and gene list}
scRNAmetaData <- read.table("~/Dropbox (Personal)/scRNA/SGE_DM/data/pseudobulk_info.txt", sep = ",", header = TRUE)

#

dmMetaData <- data.frame(c("00","01","02","03","04","05","06","07","08","09","10"))

colnames(dmMetaData) <- "clusterID"

dmMetaData$nGenes <- c(6251,7359,2907,749,2541,5416,1214,1266,390,1350,2074)

dmMetaData$nInds <- c(45,45,45,31,45,45,41,43,44,26,20)

dmMetaData$nCells <- as.vector(colSums(scRNAmetaData[,5:15]))

ggplot(dmMetaData, aes(x = nCells, y = nGenes)) +
  geom_point() +
  geom_text(aes(label = clusterID), nudge_x = 1000) +
  xlab("Number of Cells in scRNA cluster") +
  ylab("Number of Genes with non-NA DM for all Individuals")
ggsave(paste0(figDir,"genes_v_clusterCells.png"))

ggplot(dmMetaData, aes(x = nCells, y = nInds)) +
  geom_point() +
  ylim(c(0,45)) +
  geom_text(aes(label = clusterID), nudge_y = -2) +
  xlab("Number of Cells in scRNA cluster") +
  ylab("Number of Individuals with 5 or more Cells")
ggsave(paste0(figDir,"individuals_v_clusterCells.png"))
```

cluster 3 cell counts by individual
```{r}
cellCounts <- read.table("~/Dropbox (Personal)/scRNA/SGE_DM/data/NC_cells_by_ind_large_clusters", header = TRUE)

p1 <- ggplot(data = cellCounts, aes(x = c0)) +
  geom_histogram(binwidth = 50) + ggtitle("CD8_A") +
  geom_vline(xintercept = 5, col = "red", alpha = .75)

p2 <- ggplot(data = cellCounts, aes(x = c1)) +
  geom_histogram(binwidth = 50) + ggtitle("CD4") +
  geom_vline(xintercept = 5, col = "red", alpha = .75)

p3 <- ggplot(data = cellCounts, aes(x = c2)) +
  geom_histogram(binwidth = 50) + ggtitle("NK") +
  geom_vline(xintercept = 5, col = "red", alpha = .75)

p4 <- ggplot(data = cellCounts, aes(x = c3)) +
  geom_histogram(binwidth = 50) + ggtitle("CD8_B") +
  geom_vline(xintercept = 5, col = "red", alpha = .75)

p5 <- ggplot(data = cellCounts, aes(x = c4)) +
  geom_histogram(binwidth = 50) + ggtitle("CD4") +
  geom_vline(xintercept = 5, col = "red", alpha = .75)

p6 <- ggplot(data = cellCounts, aes(x = c5)) +
  geom_histogram(binwidth = 50) + ggtitle("B Cells") +
  geom_vline(xintercept = 5, col = "red", alpha = .75)

p7 <- ggplot(data = cellCounts, aes(x = c14)) +
  geom_histogram(binwidth = 50) + ggtitle("CD4 c1+c4") +
  geom_vline(xintercept = 5, col = "red", alpha = .75)

plot_grid(p1,p2,p3,p4,p6,p7, ncol = 3)
ggsave(paste0(figDir,"cellCount_by_cellType.png"))

```



```{r correlations between mean and DM}
## just take a quick look at some stuff ##
meanVarRs <- meanDMRs <- pbMeanVarRs <- pbMeanDMRs <- vector()

for (i in 1:45){
  meanVarRs[[i]] <- cor.test(sge_NC_mean_matrix[,i],sge_NC_var_matrix[,i], method = "spearman")$estimate 
}

for (i in 1:45){
  meanDMRs[[i]] <- cor.test(sge_NC_mean_matrix[,i],sge_NC_DM_matrix[,i], method = "spearman")$estimate 
}

for (i in 1:45){
  pbMeanVarRs[[i]] <- cor.test(sge_NC_pbMean_matrix[,i],sge_NC_var_matrix[,i], method = "spearman")$estimate 
}

for (i in 1:45){
  pbMeanDMRs[[i]] <- cor.test(sge_NC_pbMean_matrix[,i],sge_NC_DM_matrix[,i], method = "spearman")$estimate 
}


ggplot() +
  geom_density(aes(meanVarRs, fill = 'mean/variance')) +
  geom_density(aes(meanDMRs, fill = 'mean/DM')) +
  #xlim(-0.01,1) +
  xlab('R') +
  scale_fill_discrete(name = "")
ggsave(paste0(figDir,"c",cluster,"/","hist_mean_dm_and_var_Rvalues_c",cluster,"_spearman.png"))

ggplot() +
  geom_density(aes(pbMeanVarRs, fill = 'pbMean/variance')) +
  geom_density(aes(pbMeanDMRs, fill = 'pbMean/DM')) +
  #xlim(-0.01,1) +
  xlab('R') +
  scale_fill_discrete(name = "")
ggsave(paste0(figDir,"c",cluster,"/","hist_pbMean_dm_and_var_Rvalues_c",cluster,"_spearman.png"))

ggplot() +
  geom_density(aes(meanVarRs, fill = 'mean/variance')) +
  geom_density(aes(pbMeanVarRs, fill = 'pbMean/variance')) +
  #xlim(-0.01,1) +
  xlab('R') +
  scale_fill_discrete(name = "")
ggsave(paste0(figDir,"c",cluster,"/","hist_means_and_var_Rvalues_c",cluster,"_spearman.png"))

ggplot() +
  geom_density(aes(meanDMRs, fill = 'mean/DM')) +
  geom_density(aes(pbMeanDMRs, fill = 'pbMean/DM')) +
  #xlim(-0.01,1) +
  xlab('R') +
  scale_fill_discrete(name = "")
ggsave(paste0(figDir,"c",cluster,"/","hist_means_and_dm_Rvalues_c",cluster,"_spearman.png"))



ggplot() +
  geom_density(aes(abs(sge_NC_DM_matrix[,1]), fill='NC'), alpha=0.5) +
  geom_density(aes(abs(sge_LPS_DM_matrix[,1]), fill='LPS'), alpha=0.5) +
  xlim(0,0.4)

```


```{r which betas indicate noisy LPS genes}
clusterCellID <- as.data.frame(c("00","14","02","03","05","06","07"))

colnames(clusterCellID) <- "cluster"

clusterCellID$cellData <- c("~/Dropbox (Personal)/scRNA/SGE_DM/paulData/res_fdr_celltype_CD8_A_modelnum_1_covariate_elo_nperms_100.rds",
                            "~/Dropbox (Personal)/scRNA/SGE_DM/paulData/res_fdr_celltype_CD4_modelnum_1_covariate_elo_nperms_100.rds",
                            "~/Dropbox (Personal)/scRNA/SGE_DM/paulData/res_fdr_celltype_NK_modelnum_1_covariate_elo_nperms_100.rds",
                            "~/Dropbox (Personal)/scRNA/SGE_DM/paulData/res_fdr_celltype_CD8_B_modelnum_1_covariate_elo_nperms_100.rds",
                            "~/Dropbox (Personal)/scRNA/SGE_DM/paulData/res_fdr_celltype_B_cells_modelnum_1_covariate_elo_nperms_100.rds",
                            "~/Dropbox (Personal)/scRNA/SGE_DM/paulData/res_fdr_celltype_mono.CD14hi_modelnum_1_covariate_elo_nperms_100.RDS",
                            "~/Dropbox (Personal)/scRNA/SGE_DM/paulData/res_fdr_celltype_mono.CD14low_modelnum_1_covariate_elo_nperms_100.RDS")

clusterCellID$cellID <- c("CD8A",
                          "CD4",
                          "NK",
                          "CD8B",
                          "Bcells",
                          "CD14hi",
                          "CD14low")

#00 = CD8_A
#01 = cd4
#02 = NK
#03 = CD8_B
#04 = cd4 (1&4 combined in paul's analysis)
#05 = B
#06 = CD14hi
#07 = CD14lo
#14 = CD4 1 & 4 combined

for (c in 1:dim(clusterCellID)[1]) {
  cluster <- clusterCellID$cluster[c]
  paulResults <- readRDS(clusterCellID$cellData[c])
  cellID <- clusterCellID$cellID[c]
  
  DMPBemmreml <- get(paste0("res_full_DMpb_c",cluster))
  
  paulResults <- paulResults[rownames(paulResults) %in% geneList,]
  
  expData <- paulResults[,c(2,10,18)]
  expData$stdbeta_trtLPS <- expData$beta_trtLPS / sqrt(expData$varbeta_trtLPS)
  colnames(expData) <- paste0(colnames(expData),"_exp")
  expData$gene <- rownames(expData)
  
  dmpbData <- DMPBemmreml[,c(2,8,14)]
  dmpbData$stdbeta_trtNC <- dmpbData$beta_trtNC / sqrt(dmpbData$sdev_trtNC)
  colnames(dmpbData) <- paste0(colnames(dmpbData),"_dmpb")
  dmpbData$gene <- rownames(dmpbData)
  
  
  #c01betas <- merge(expData, dmData, by = "gene")
  c01betas <- merge(expData, dmpbData, by = "gene")
  
  rownames(c01betas) <- c01betas$gene
  
  #how do the betas from treatment look across the 3 models
  
  # ggplot(data = c01betas, aes(x = stdbeta_trtNC_dm, y = stdbeta_trtNC_dmpb)) +
  #   geom_point(alpha = .1) +
  #   xlab("Standardized Treatment Betas - DM model") +
  #   ylab("Standardized Treatment Betas - DMpb model") +
  #   ggtitle(paste0("cluster ",cluster))
  # ggsave(paste0(figDir,"c",cluster,"/","StdBetas_DMpb_v_DM_c",cluster,".png"), width = 6, height = 4)
  
  ggplot(data = c01betas, aes(x = stdbeta_trtLPS_exp, y = stdbeta_trtNC_dmpb)) +
    geom_point(alpha = .1) +
    xlab("Standardized Treatment Betas - Pesudobulk model (from Paul)") +
    ylab("Standardized Treatment Betas - DMpb model") +
    ggtitle(paste0("cluster ",cluster))
  ggsave(paste0(figDir,"c",cluster,"/","StdBetas_DMpb_v_PaulBetas_c",cluster,".png"), width = 6, height = 4)
  
  
  dmpbCutoff <- quantile(c01betas$stdbeta_trtNC_dmpb, c(.05))
  
  noisyLPSgenes <- subset(c01betas, stdbeta_trtNC_dmpb < dmpbCutoff & stdbeta_trtLPS_exp > 0)
  nNoisy <- dim(noisyLPSgenes)[1]
  perNoisy <- 100 * round(nNoisy / dim(c01betas)[1], digits = 3)
  
  ggplot() +
    geom_point(data = c01betas, aes(x = stdbeta_trtLPS_exp, y = stdbeta_trtNC_dmpb), alpha = .1) +
    geom_point(data = noisyLPSgenes, aes(x = stdbeta_trtLPS_exp, y = stdbeta_trtNC_dmpb), col = "red", alpha = .5) +
    xlab("Standardized Treatment Betas - Pesudobulk model (from Paul)") +
    ylab("Standardized Treatment Betas - DMpb model") +
    ggtitle(paste0("cluster ",cluster," ",cellID," cells"),
            subtitle = paste0(nNoisy," highlighted genes, or ",perNoisy,"%"))
  ggsave(paste0(figDir,"c",cluster,"/","StdBetas_DMpb_v_PaulBetas_noisyGenes_c",cluster,".png"), width = 6, height = 4)
  
  sge_LPS_pbMean_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c',cluster,'_pbMean_matrix'))
  sge_LPS_DMpb_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c',cluster,'_DMpb_matrix'))
  sge_LPS_var_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_LPS_c',cluster,'_var_matrix'))
  
  sge_NC_pbMean_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c',cluster,'_pbMean_matrix'))
  sge_NC_DMpb_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c',cluster,'_DMpb_matrix'))
  sge_NC_var_matrix <- read.table(paste0('~/Dropbox (Personal)/scRNA/SGE_DM/data/sge_NC_c',cluster,'_var_matrix'))
  
  #don't need all of them, just plot the top 5
  #for (geneName in rownames(noisyLPSgenes)) {
  for (geneName in rownames(head(noisyLPSgenes[order(noisyLPSgenes$stdbeta_trtNC_dmpb),], n = 5))) {
    #geneName <- "ADAM19"
    lpsDMpb <- sge_LPS_DMpb_matrix[which(rownames(sge_LPS_DMpb_matrix) == geneName),]
    ncDMpb <- sge_NC_DMpb_matrix[which(rownames(sge_NC_DMpb_matrix) == geneName),]
    
    lpsPBmean <- sge_LPS_pbMean_matrix[which(rownames(sge_LPS_pbMean_matrix) == geneName),]
    ncPBmean <- sge_NC_pbMean_matrix[which(rownames(sge_NC_pbMean_matrix) == geneName),]
    
    geneData <- t(rbind(lpsDMpb,lpsPBmean,ncDMpb,ncPBmean))
    colnames(geneData) <- c("lpsDMpb","lpsPBmean","ncDMpb","ncPBmean")
    geneData <- as.data.frame(geneData)
    
    DMpbRaw<- melt(geneData[,c(1,3)])
    DMpbRaw$ID <- c(rownames(geneData),rownames(geneData))
    DMpbRaw$variable <- factor(DMpbRaw$variable, levels = c("ncDMpb", "lpsDMpb"))
    
    PBmeanRaw<- melt(geneData[,c(2,4)])
    PBmeanRaw$ID <- c(rownames(geneData),rownames(geneData))
    PBmeanRaw$variable <- factor(PBmeanRaw$variable, levels = c("ncPBmean", "lpsPBmean"))
    
    ggplot(data = PBmeanRaw, aes(x = variable, y = value, group = ID, col = ID)) +
      geom_path() + theme(legend.position = "none") +
      xlab("treatment") + ylab("pseudobulk mean") +
      ggtitle(paste0("cluster ",cluster,", cell type ",cellID,", gene ",geneName))
    ggsave(paste0(figDir,"c",cluster,"/","noisyGenes_raw_pbMean_c",cluster,"_gene_",geneName,".png"), width = 6, height = 4)
    
    ggplot(data = DMpbRaw, aes(x = variable, y = value, group = ID, col = ID)) +
      geom_path() + theme(legend.position = "none") +
      xlab("treatment") + ylab("DM from pseudobulk") +
      ggtitle(paste0("cluster ",cluster,", cell type ",cellID,", gene ",geneName))
    ggsave(paste0(figDir,"c",cluster,"/","noisyGenes_raw_DMpb_c",cluster,"_gene_",geneName,".png"), width = 6, height = 4)
      
    ggplot() +
      geom_point(data = geneData, aes(x = lpsPBmean, y = lpsDMpb), col = "red") +
      geom_point(data = geneData, aes(x = ncPBmean, y = ncDMpb), col = "gray") +
      xlab("pseudobulk mean") + ylab("DM from pseudobulk") +
      ggtitle(paste0("cluster ",cluster,", cell type ",cellID,", gene ",geneName),
              subtitle = "red = LPS, grey = NC")
    ggsave(paste0(figDir,"c",cluster,"/","noisyGenes_raw_DMpb_v_pbMean_c",cluster,"_gene_",geneName,".png"), width = 6, height = 4)
  }
  
  #rerun this to grab two genes that are high LPS but no change in DM
  # then check those genes for LPS v. var - basically does high LPS always produce high variance?
  pbMeanCutoff <- quantile(c01betas$stdbeta_trtLPS_exp, c(.75))
  highLPSgenes <- subset(c01betas, stdbeta_trtLPS_exp > pbMeanCutoff & abs(stdbeta_trtNC_dmpb) < .5)
  nHigh <- dim(highLPSgenes)[1]
  
  ggplot() +
    geom_point(data = c01betas, aes(x = stdbeta_trtLPS_exp, y = stdbeta_trtNC_dmpb), alpha = .1) +
    geom_point(data = highLPSgenes, aes(x = stdbeta_trtLPS_exp, y = stdbeta_trtNC_dmpb), col = "blue", alpha = .5) +
    xlab("Standardized Treatment Betas - Pesudobulk model (from Paul)") +
    ylab("Standardized Treatment Betas - DMpb model") +
    ggtitle(paste0("cluster ",cluster," ",cellID," cells"),
            subtitle = paste0(nHigh," highlighted genes"))
  ggsave(paste0(figDir,"c",cluster,"/","StdBetas_DMpb_v_PaulBetas_highLPSGenes_c",cluster,".png"), width = 6, height = 4)
  
  
  #don't need all of them, just pick 5 at random
  randomGeneNames <- sample(rownames(highLPSgenes), size = 5, replace = F)
  #for (geneName in rownames(noisyLPSgenes)) {
  
  highLPSrandomPlot <- highLPSgenes[rownames(highLPSgenes) %in% randomGeneNames,]

  ggplot() +
    geom_point(data = c01betas, aes(x = stdbeta_trtLPS_exp, y = stdbeta_trtNC_dmpb), alpha = .1) +
    geom_point(data = highLPSrandomPlot, aes(x = stdbeta_trtLPS_exp, y = stdbeta_trtNC_dmpb), col = "blue", alpha = .5) +
    xlab("Standardized Treatment Betas - Pesudobulk model (from Paul)") +
    ylab("Standardized Treatment Betas - DMpb model") +
    ggtitle(paste0("cluster ",cluster," ",cellID," cells"),
            subtitle = paste0(nHigh," highlighted genes"))
  ggsave(paste0(figDir,"c",cluster,"/","StdBetas_DMpb_v_PaulBetas_highLPSGenes_random_c",cluster,".png"), width = 6, height = 4)
  
  
  
  for (geneName in randomGeneNames) {
    #geneName <- sample(rownames(highLPSgenes))[1]
    
    lpsDMpb <- sge_LPS_DMpb_matrix[which(rownames(sge_LPS_DMpb_matrix) == geneName),]
    ncDMpb <- sge_NC_DMpb_matrix[which(rownames(sge_NC_DMpb_matrix) == geneName),]
    
    lpsPBmean <- sge_LPS_pbMean_matrix[which(rownames(sge_LPS_pbMean_matrix) == geneName),]
    ncPBmean <- sge_NC_pbMean_matrix[which(rownames(sge_NC_pbMean_matrix) == geneName),]

    lpsVar <- sge_LPS_var_matrix[which(rownames(sge_LPS_var_matrix) == geneName),]
    ncVar <- sge_NC_var_matrix[which(rownames(sge_NC_var_matrix) == geneName),]
    
    geneData <- t(rbind(lpsDMpb,lpsPBmean,lpsVar,ncDMpb,ncPBmean,ncVar))
    colnames(geneData) <- c("lpsDMpb","lpsPBmean","lpsVar","ncDMpb","ncPBmean","ncVar")
    geneData <- as.data.frame(geneData)
    
    DMpbRaw <- melt(geneData[,c(1,4)])
    DMpbRaw$ID <- c(rownames(geneData),rownames(geneData))
    DMpbRaw$variable <- factor(DMpbRaw$variable, levels = c("ncDMpb", "lpsDMpb"))
    
    PBmeanRaw <- melt(geneData[,c(2,5)])
    PBmeanRaw$ID <- c(rownames(geneData),rownames(geneData))
    PBmeanRaw$variable <- factor(PBmeanRaw$variable, levels = c("ncPBmean", "lpsPBmean"))
    
    VarRaw <- melt(geneData[,c(3,6)])
    VarRaw$ID <- c(rownames(geneData),rownames(geneData))
    VarRaw$variable <- factor(VarRaw$variable, levels = c("ncVar", "lpsVar"))

    
    ggplot(data = PBmeanRaw, aes(x = variable, y = value, group = ID, col = ID)) +
      geom_path() + theme(legend.position = "none") +
      xlab("treatment") + ylab("pseudobulk mean") +
      ggtitle(paste0("cluster ",cluster,", cell type ",cellID,", gene ",geneName))
    ggsave(paste0(figDir,"c",cluster,"/","highLPSGenes_raw_pbMean_c",cluster,"_gene_",geneName,".png"), width = 6, height = 4)
    
    ggplot(data = DMpbRaw, aes(x = variable, y = value, group = ID, col = ID)) +
      geom_path() + theme(legend.position = "none") +
      xlab("treatment") + ylab("DM from pseudobulk") +
      ggtitle(paste0("cluster ",cluster,", cell type ",cellID,", gene ",geneName))
    ggsave(paste0(figDir,"c",cluster,"/","highLPSGenes_raw_DMpb_c",cluster,"_gene_",geneName,".png"), width = 6, height = 4)

    ggplot(data = VarRaw, aes(x = variable, y = value, group = ID, col = ID)) +
      geom_path() + theme(legend.position = "none") +
      xlab("treatment") + ylab("raw Variance") +
      ggtitle(paste0("cluster ",cluster,", cell type ",cellID,", gene ",geneName))
    ggsave(paste0(figDir,"c",cluster,"/","highLPSGenes_raw_Var_c",cluster,"_gene_",geneName,".png"), width = 6, height = 4)
    
    ggplot() +
      geom_point(data = geneData, aes(x = lpsPBmean, y = lpsDMpb), col = "red") +
      geom_point(data = geneData, aes(x = ncPBmean, y = ncDMpb), col = "gray") +
      xlab("pseudobulk mean") + ylab("DM from pseudobulk") +
      ggtitle(paste0("cluster ",cluster,", cell type ",cellID,", gene ",geneName),
              subtitle = "red = LPS, grey = NC")
    ggsave(paste0(figDir,"c",cluster,"/","highLPSGenes_raw_DMpb_v_pbMean_c",cluster,"_gene_",geneName,".png"), width = 6, height = 4)
    
    ggplot() +
      geom_point(data = geneData, aes(x = lpsPBmean, y = lpsVar), col = "red") +
      geom_point(data = geneData, aes(x = ncPBmean, y = ncVar), col = "gray") +
      xlab("pseudobulk mean") + ylab("raw Variance") +
      ggtitle(paste0("cluster ",cluster,", cell type ",cellID,", gene ",geneName),
              subtitle = "red = LPS, grey = NC")
    ggsave(paste0(figDir,"c",cluster,"/","highLPSGenes_raw_Var_v_pbMean_c",cluster,"_gene_",geneName,".png"), width = 6, height = 4)

  }
  
}


```


```{r how do the mean betas compare to the DM betas}

ggplot(data = c01betas, aes(x = stdbeta_trtLPS_exp, y = stdbeta_trtNC_dm)) +
  geom_point(alpha = .1)

#what about the betas from elo:trt
dmDataELO <- res_fullDFc01[,c(2,8,14)+1]
dmDataELO$stdbeta_trtLPS_elo_centered <- dmDataELO$beta_trtLPS.elo_centered / sqrt(dmDataELO$sdev_trtLPS.elo_centered)
colnames(dmDataELO) <- paste0(colnames(dmDataELO),"_dm")
dmDataELO$gene <- rownames(dmDataELO)


dmpbDataELO <- res_full_DMpb_c01[,c(2,8,14)+1]
dmpbDataELO$stdbeta_trtLPS_elo_centered <- dmpbDataELO$beta_trtLPS.elo_centered / sqrt(dmpbDataELO$sdev_trtLPS.elo_centered)
colnames(dmpbDataELO) <- paste0(colnames(dmpbDataELO),"_dmpb")
dmpbDataELO$gene <- rownames(dmpbDataELO)

c01betasELO <- merge(dmDataELO, dmpbDataELO, by = "gene")

ggplot(data = c01betasELO, aes(x = stdbeta_trtLPS_elo_centered_dm, y = stdbeta_trtLPS_elo_centered_dmpb)) +
  geom_point(alpha = .1)

```

now that we've settled on a model, how many genes pass a basic qval
```{r}
#in clusters 00 -> 07
#how many pass fdr of .01, .05, .10
#treatment
#LPS:elo

FDRs <- c(.01,.05,.1,.2)

trtFDR <- as.data.frame(clusterCellID$cluster)
elotrtFDR <- as.data.frame(clusterCellID$cluster)
agAsymtrtFDR <- as.data.frame(clusterCellID$cluster)
agRectrtFDR <- as.data.frame(clusterCellID$cluster)
groomtrtFDR <- as.data.frame(clusterCellID$cluster)

colnames(trtFDR) <- colnames(elotrtFDR) <- colnames(agAsymtrtFDR) <- colnames(agRectrtFDR) <- colnames(groomtrtFDR) <- "cluster"

trtFDR$nGenes <- trtFDR$fdr20 <- trtFDR$fdr10 <- trtFDR$fdr05 <- trtFDR$fdr01 <- 0
elotrtFDR$nGenes <- elotrtFDR$fdr20 <- elotrtFDR$fdr10 <- elotrtFDR$fdr05 <- elotrtFDR$fdr01 <- 0
agAsymtrtFDR$nGenes <- agAsymtrtFDR$fdr20 <- agAsymtrtFDR$fdr10 <- agAsymtrtFDR$fdr05 <- agAsymtrtFDR$fdr01 <- 0
agRectrtFDR$nGenes <- agRectrtFDR$fdr20 <- agRectrtFDR$fdr10 <- agRectrtFDR$fdr05 <- agRectrtFDR$fdr01 <- 0
groomtrtFDR$nGenes <- groomtrtFDR$fdr20 <- groomtrtFDR$fdr10 <- groomtrtFDR$fdr05 <- groomtrtFDR$fdr01 <- 0


for (c in 1:dim(clusterCellID)[1]) {
  cluster <- clusterCellID$cluster[c]
  
  cellID <- clusterCellID$cellID[c]
  
  for (f in FDRs) {
    DMPBemmreml <- get(paste0("res_full_DMpb_c",cluster))
    trtFDR[c,(which(FDRs == f) + 1)] <- sum(qvalue(p = DMPBemmreml$p_value_trtNC, fdr.level = f)$significant)
    elotrtFDR[c,(which(FDRs == f) + 1)] <- sum(qvalue(p = DMPBemmreml$p_value_trtLPS.elo_centered, fdr.level = f)$significant)
    trtFDR$nGenes[c] <- dim(DMPBemmreml)[1]
    elotrtFDR$nGenes[c] <- dim(DMPBemmreml)[1]
    
    DMPBemmreml <- get(paste0("res_full_aggAsym_DMpb_c",cluster))
    agAsymtrtFDR[c,(which(FDRs == f) + 1)] <- sum(qvalue(p = DMPBemmreml$p_value_trtLPS.aggAsymm_cent, fdr.level = f)$significant)
    agAsymtrtFDR$nGenes[c] <- dim(DMPBemmreml)[1]
    
    DMPBemmreml <- get(paste0("res_full_aggRec_DMpb_c",cluster))
    agRectrtFDR[c,(which(FDRs == f) + 1)] <- sum(qvalue(p = DMPBemmreml$p_value_trtLPS.aggRec_cent, fdr.level = f)$significant)
    agRectrtFDR$nGenes[c] <- dim(DMPBemmreml)[1]
    
    DMPBemmreml <- get(paste0("res_full_groom_DMpb_c",cluster))
    groomtrtFDR[c,(which(FDRs == f) + 1)] <- sum(qvalue(p = DMPBemmreml$p_value_trtLPS.groom_cent, fdr.level = f)$significant)
    groomtrtFDR$nGenes[c] <- dim(DMPBemmreml)[1]
  }
  
}


write.table(trtFDR, file = paste0(workDir,"treatmentFDRcounts.txt"), quote = F, sep = ",")
write.table(elotrtFDR, file = paste0(workDir,"eloTrtFDRcounts.txt"), quote = F, sep = ",")
write.table(agAsymtrtFDR, file = paste0(workDir,"agAsymTrtFDRcounts.txt"), quote = F, sep = ",")
write.table(agRectrtFDR, file = paste0(workDir,"agRecTrtFDRcounts.txt"), quote = F, sep = ",")
write.table(groomtrtFDR, file = paste0(workDir,"groomTrtFDRcounts.txt"), quote = F, sep = ",")

```


```{r pvalue plots}
#elo
for (c in 1:dim(clusterCellID)[1]) {
  cluster <- clusterCellID$cluster[c]
  DMPBemmreml <- get(paste0("res_full_DMpb_c",cluster))

  plot_grid(ggplot(data = DMPBemmreml, aes(x = p_value_.Intercept.)) +
    geom_histogram(bins = 100) + ggtitle(paste0("elo_DMpb_c",cluster)),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtNC.age_centered)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtNC.elo_centered)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtNC)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtLPS.age_centered)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtLPS.elo_centered)) +
    geom_histogram(bins = 100), ncols = 3)
  ggsave(paste0(figDir,"c",cluster,"/elo_pval_allHists_",cluster,".png"), width = 8, height = 6)
}

#agRec
for (c in 1:dim(clusterCellID)[1]) {
  cluster <- clusterCellID$cluster[c]
  DMPBemmreml <- get(paste0("res_full_aggRec_DMpb_c",cluster))

  plot_grid(ggplot(data = DMPBemmreml, aes(x = p_value_.Intercept.)) +
    geom_histogram(bins = 100) + ggtitle(paste0("agRec_DMpb_c",cluster)),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtNC.age_centered)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtNC.aggRec_cent)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtNC)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtLPS.age_centered)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtLPS.aggRec_cent)) +
    geom_histogram(bins = 100), ncols = 3)
  ggsave(paste0(figDir,"c",cluster,"/agRec_pval_allHists_",cluster,".png"), width = 8, height = 6)
}

#agAsym
for (c in 1:dim(clusterCellID)[1]) {
  cluster <- clusterCellID$cluster[c]
  DMPBemmreml <- get(paste0("res_full_aggAsym_DMpb_c",cluster))

  plot_grid(ggplot(data = DMPBemmreml, aes(x = p_value_.Intercept.)) +
    geom_histogram(bins = 100) + ggtitle(paste0("agAsym_DMpb_c",cluster)),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtNC.age_centered)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtNC.aggAsymm_cent)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtNC)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtLPS.age_centered)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtLPS.aggAsymm_cent)) +
    geom_histogram(bins = 100), ncols = 3)
  ggsave(paste0(figDir,"c",cluster,"/agAsym_pval_allHists_",cluster,".png"), width = 8, height = 6)
}

#groom
for (c in 1:dim(clusterCellID)[1]) {
  cluster <- clusterCellID$cluster[c]
  DMPBemmreml <- get(paste0("res_full_groom_DMpb_c",cluster))

  plot_grid(ggplot(data = DMPBemmreml, aes(x = p_value_.Intercept.)) +
    geom_histogram(bins = 100) + ggtitle(paste0("groom_DMpb_c",cluster)),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtNC.age_centered)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtNC.groom_cent)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtNC)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtLPS.age_centered)) +
    geom_histogram(bins = 100),
    ggplot(data = DMPBemmreml, aes(x = p_value_trtLPS.groom_cent)) +
    geom_histogram(bins = 100), ncols = 3)
  ggsave(paste0(figDir,"c",cluster,"/grooming_pval_allHists_",cluster,".png"), width = 8, height = 6)
}
       
       
       
```

GSEA for Agonisms Received (strongest signal)
```{r GSEA and pathway analysis}
### Permutation-based GSEA tests
### Originally developed by Jordan Anderson and Noah Snyder-Mackler. Modified by Tauras Vilgalys. 
### Last updated 2020 September 16

###outside R, build a csv file with geneName,hallmarkPathway

# Parameter options on lines 11 and 12
# Data and files to load on lines 16 and 30
## parameters to consider: 
pvals <- F ## are data in pvalues or betas? If betas, set pvals to FALSE
weight <- 1 ## Weight to place on betas. Typically either a 1 or a 0, where 0 weights all effects equally.
        ## weight=0 acts like a KS test, and is typically used with pvalues. 

sampleBetas <- as.data.frame(rownames(fluEXPemmreml))
sampleBetas$Betas <- fluEXPemmreml$beta_treatment

colnames(sampleBetas) <- c("Gene","Betas")

## get gene.list
gene.list <- sampleBetas$Gene

## get nperm permutations of the vector of Betas
nperm <- 250
for (i in 1:nperm) {
  tmpB <- sample(sampleBetas$Beta)
  cbind(sampleBetas,tmpB) -> sampleBetas; rm(tmpB)
}; rm(i)#; head(sampleBetas)

## Get the list of gene sets to test. here the file starts with gene names and the associated GO terms. 

## i downloaded these hallmark gene sets from https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp

gene.set <- read.table("~/Dropbox (Personal)/DNHS/hallmark_genes/hallmarkPathways.csv", sep = ",")

colnames(gene.set) <- c("Associated.Gene.Name","GO.Term.Accession")

gene.set$gene_GO <- paste0(gene.set$Associated.Gene.Name,"_",gene.set$GO.Term.Accession)

# Trim for expressed genes where B was estimated
gene.set2 <- subset(gene.set, gene.set$Associated.Gene.Name %in% sampleBetas$Gene)


# Remove duplicated gene/GO pairings
gene.set2 <- gene.set2[!duplicated(gene.set2$gene_GO),]

## Get list of gene ontologies to test, and the number of genes in each
GO_res <- as.data.frame(matrix(ncol=2, nrow=length(unique(gene.set2$GO.Term.Accession)), unique(gene.set2$GO.Term.Accession)))
GO_res$V2 <- 0
GO_res <- GO_res[!(GO_res$V1 == ""),]; colnames(GO_res) <- c("GO", "n_genes")
for (i in 1:nrow(GO_res)) {
  GO_res$n_genes[i] <- sum(gene.set2$GO.Term.Accession == GO_res$GO[i])
  }
rm(i)
GO_res$n_genes <- as.numeric(GO_res$n_genes)

## Only consider gene sets with at least 5 measured genes
GO_res <- subset(GO_res, n_genes >= 5)

## Create a matrix of just the effect sizes, moving the gene names to the row.names
effect_size_matrix<-sampleBetas[,-1]
row.names(effect_size_matrix) <- sampleBetas[,1]


## GSEA function
### p-values are calculated based on a cdf. Another option would be to use a z-statistic, however that would assume normality? 
GSEA=function(effect_size_matrix,gene_set_list,gene_GO_associations, weight=1,num.cores=NA,pvals=F){
  ## effect_size_matrix == gene x permutations matrix where rownames are the genes and the first column is the observed effect sizes (i.e., columns 2:ncol are the permuted betas)
  if (is.na(num.cores)) num.cores=detectCores()
  clus = makeCluster(getOption("cl.cores", num.cores))
  clusterExport(clus,varlist=ls(),envir=environment())
  a=parApply(clus,effect_size_matrix,2,function(tmp_data){
    names(tmp_data)=rownames(effect_size_matrix)
    tmp_data=sort(tmp_data,decreasing = T)
    correl.vector<-tmp_data
    tmp_all_genes<-names(tmp_data)
    #So that we ultimately weight on our betas when weight==1
    weighted.score.type=weight
    Reduce(rbind,lapply(gene_set_list,function(GOcat){
      tmp_set <- gene_GO_associations$Associated.Gene.Name[gene_GO_associations$GO.Term.Accession == GOcat]
      #First we're indexing what genes are or aren't in our gene set (1 vs 0)
      tag.indicator <- sign(match(tmp_all_genes, tmp_set, nomatch=0)) 
      no.tag.indicator <- 1 - tag.indicator 
      N <- length(tmp_all_genes) 
      Nh <- length(tmp_set) 
      Nm <-  N - Nh 
      #This is just if we want to do a KS test(maybe if we want to compare later to some of your results using GAGE); weight==0
      if (weight == 0) {
        correl.vector <- rep(1, N)
      }
      alpha <- weighted.score.type
      if (pvals==T) {correl.vector=-log10(correl.vector)}
      correl.vector <- abs(correl.vector**alpha)
      #Normalizing our Incremental increase by the sum of the abs our "correlations" (sum.correl.tag) and 
      #normalizing our incremental decrease by the number of genes not in our set (Nm, which remember came from N-Nh above).
      sum.correl.tag  <- sum(correl.vector[which(tag.indicator == 1)])
      #attempting to remove the normalization
      norm.tag    <- 1.0 / sum.correl.tag
      norm.no.tag <- 1.0 / Nm
      #Then we just walk down the gene list adding or substracting as such
      ## don't use normalization in the math
      RES <- cumsum(tag.indicator * correl.vector * norm.tag - no.tag.indicator * norm.no.tag)
      #RES <- cumsum(tag.indicator * correl.vector - no.tag.indicator)
      #Then obviously this is for maximum positively vs negatively enriched 
      max.ES <- max(RES)
      min.ES <- min(RES)
      if (max.ES > - min.ES) {
        #      ES <- max.ES
        ES <- signif(max.ES, digits = 5)
        arg.ES <- which.max(RES)
      } else {
        #      ES <- min.ES
        ES <- signif(min.ES, digits=5)
        arg.ES <- which.min(RES)
      }
      #if (length(gene_set)==1) {plot(cumsum(tag.indicator * correl.vector * norm.tag - no.tag.indicator * norm.no.tag),type = "l",xlab="Gene Index",ylab="Running Enrichment Score",xaxt="n"); axis(1,at=tag.indicator*1:length(tag.indicator),labels=T,tick=T)}
      return(c(max.ES,min.ES))}))
  })
  ## Returns two columns, one for upreg and one for downreg
  a=data.frame(a)
  a_upreg=a[1:length(gene_set_list),]
  a_downreg=a[(length(gene_set_list)+1):nrow(a),]
  rownames(a_upreg)=names(gene_set_list)
  rownames(a_downreg)=names(gene_set_list)
  #rownames(a)=names(gene.set)
  ## This doesn't seem right. Why would the pvals be lower if the max doesn't exceed the null?
  pvals_upreg=apply(a_upreg,1,function(x){
    #signif((sum(x[1] >= x[2:length(x)])+1)/(length(x)), digits=5)
    ee <- ecdf(x[2:length(x)]); 1 - ee(x[1])
    })
  names(pvals_upreg)=names(gene_set_list)
  pvals_downreg=apply(a_downreg,1,function(x){
    #signif((sum(x[1] <= x[2:length(x)])+1)/(length(x)), digits=5)
    ee <- ecdf(x[2:length(x)]); ee(x[1])
    })
  names(pvals_downreg)=names(gene_set_list)
  pvals=cbind(pvals_upreg,pvals_downreg); colnames(pvals)=c("pval_upreg","pval_downreg")
  output=list(a_upreg,a_downreg,pvals)
}

## Run GSEA function for each GO term (defined earlier in GO_res$GO)
output<-GSEA(effect_size_matrix = effect_size_matrix,
             gene_set_list = unique(GO_res$GO),
             gene_GO_associations = gene.set2,
             weight = 1, pvals=F)

names(output)=c("NES_matrix_upreg","NES_matrix_downreg","p_values")




# Normalize + ES' by the mean of + ES values, and - ES' by the mean of - ES null values for that gene set. 
##This should effectively control for differences in gene set size
pos.ES.norm<-matrix(0,nrow=nrow(GO_res),ncol=nperm+1)
neg.ES.norm<-matrix(0,nrow=nrow(GO_res),ncol=nperm+1)
for (i in 1:length(GO_res$GO)) {
         pos.ES <- unlist(output$NES_matrix_upreg[i,])
         names(pos.ES)<-NULL
         neg.ES <- unlist(output$NES_matrix_downreg[i,])
          names(neg.ES)<-NULL
         #for a given gene set, N, calculate the mean of the positive max ES for estimated+ permuted ES.
         #calculate the mean of the max negative ES for estimate+ permuted ES also
         pos.m <- mean(pos.ES)
         neg.m <- mean(abs(neg.ES))
         #Then for each permutation in that geneset, scale ES by this mean value, depending on if it was max + or max - ES
         for (j in 1:(nperm+1)){
         pos.ES.norm[i,j]<-pos.ES[j]/pos.m
         neg.ES.norm[i,j]<-neg.ES[j]/neg.m
         }
}
# FDR? 
## Calculate p-values for one of the null distributions as well
perm_upreg <- apply(output$NES_matrix_upreg[,-1],1,function(x) {ee <- ecdf(x[2:length(x)]); 1 - ee(x[1])})
perm_downreg <- apply(output$NES_matrix_downreg[,-1],1,function(x) {ee <- ecdf(x[2:length(x)]); 1 - ee(x[1])})
## Use perm.fdr function (written by Joaquin Sanz, available from Snyder-Mackler et al. 2016: https://github.com/nsmackler/status_genome_2016/blob/master/perm_FDR.R)


fdr_up <- perm.fdr(output$p_values, perm_upreg, "pval_upreg", "fdr_vsPerm1")
fdr_down <- perm.fdr(output$p_values, perm_upreg, "pval_downreg", "fdr_vsPerm1")

upregNES <- output$NES_matrix_upreg[,1]
dnregNES <- output$NES_matrix_downreg[,1]


cbind(as.character(GO_res$GO),output$p_values)

#plot significant effects
betasOnly <- as.data.frame(effect_size_matrix$Betas)
colnames(betasOnly) <- c("betas")
betasOnly$rank <- rank(betasOnly$betas)

for (r in 1:2) {
  if(sum(output$p_values[,r] < .05) > 0) {
    for (g in which(output$p_values[,r] < .05)) {
      pathwayName <- as.character(GO_res$GO)[g]
      pathwayGenes <- subset(gene.set2, GO.Term.Accession == pathwayName)$Associated.Gene.Name
      betasOnly$pathway <- FALSE
      betasOnly$pathway <- rownames(effect_size_matrix) %in% pathwayGenes
      
      betasPathway <- subset(betasOnly, pathway)
      ggplot() +
        geom_linerange(aes(x = betasOnly$rank, ymax = betasOnly$betas, ymin = 0), alpha = 66/dim(effect_size_matrix)[1]) +
        geom_linerange(aes(x = betasPathway$rank, ymax = betasPathway$betas, ymin = 0)) +
        xlab("Rank") + ggtitle(pathwayName, 
                               subtitle = paste0(subset(GO_res, GO == pathwayName)$n_genes," genes, ",
                                                 "p = ",as.numeric(output$p_values[g,r]),"    ",colnames(output$p_values)[r]))
      ggsave(paste0(workDir,"figures/gsea/gsea_stickPlot_fluEXPmodel_treatment_",pathwayName,".png"), width = 6, height = 4)
    }
  }
}

fluEXPgseaTreatment <- output

cbind(as.character(GO_res$GO),fluEXPgseaTreatment$p_values,fluEXPgseaTreatment$NES_matrix_upreg[,1],fluEXPgseaTreatment$NES_matrix_downreg[,1])



```


OLD CODE
```{r raw variance in emmreml }
  ######data QC - how do significant genes from the betas of var (control for mean) model compare to DM?
  ######did fixing the DM values to 0 alter that?
  
  ## Here I'm doing some modeling to ask what are the effects of treatment and rank on variance ##
  ## get some metadata ##

  ##### AND ADD THAT TO METADATA
  metadata$means <- colMeans(sge_mean_matrix_nested)
  
  ##### THEN I CAN CONTROL FOR THE MEAN WITH THE VARIANCES AS THE FOCAL VARIABLE IN PLACE OF DM
  ## regress out group effects from DM ##
  design <- model.matrix(~metadata$means + metadata$group)
  fit <-lmFit(sge_var_matrix_nested,design)
  intercepts=data.frame(eBayes(fit))[,1]
  resid_varMean=apply(residuals.MArrayLM(object=fit, sge_var_matrix_nested),2,function(x){x+intercepts})
  
  design = model.matrix(~trt+trt:elo_centered+trt:age_centered,data=metadata)
  
  #Declare object res_full to store in it the model results: beta coefficients, standard deviations and p values
  res_full=resid_varMean[,1:(3*ncol(design))]
  colnames(res_full)[1:ncol(design)]=paste0("beta_",colnames(design))
  colnames(res_full)[(ncol(design)+1):(2*ncol(design))]=paste0("sdev_",colnames(design))
  colnames(res_full)[((2*ncol(design))+1):(3*ncol(design))]=paste0("p_value_",colnames(design))
  
  #Declare object random_effects to store in it the individual-wise u random effects
  random_effects=resid_varMean[,1:length(indIDs)]
  colnames(random_effects)=metadata[1:length(indIDs),'ID']
  
  #Define matrix Z describing the sample-to-individual mapping
  K <- read.table('~/Dropbox (Personal)/scRNA/SGE_DM/data/SGE2kinshipmatrix.txt')
  # correcting capitalization so it matches
  rownames(K)[34] <- 'JE11'
  colnames(K)[34] <- 'JE11'
  K=K[unique(metadata$ID),unique(metadata$ID)]
  
  Z=matrix(rep(0,nrow(metadata)*ncol(K)),nrow=nrow(metadata),ncol=ncol(K))
  rownames(Z)=metadata$ID
  colnames(Z)=colnames(K)
  for(i in 1:ncol(Z)){
    set=which(metadata$ID == colnames(Z)[i])
    Z[set,i]=1
  }
  
  #emmreml
  for(i in 1:nrow(resid_varMean))
  {
    emma=emmreml(y=resid_varMean[i,],X=design,Z=as.matrix(Z),K=as.matrix(K),varbetahat=T,varuhat=T,PEVuhat=T,test=T)
    random_effects[i,]=t(emma$uhat)
    res_full[i,]=t(c(emma$betahat,emma$varbetahat,emma$pvalbeta[,"none"]))
  }
  
  assign(value = data.frame(res_full), x = paste0("res_fullDF_var_c",cluster))
  plotModelHists(get(paste0("res_fullDF_var_c",cluster)), folderName = paste0(figDir,"c",cluster,"/"), modelName = paste0("Var_model_c",cluster))
  
  
  p1 <- ggplot(sge_mean_matrix_nested, aes(sge_mean_matrix_nested[,1],resid_varMean[,1])) +
    geom_point(alpha=0.5) +
    #ylim(-500,500) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    xlab('mean gene expression - individual 1') +
    ylab('variance Resid in gene expression - individual 1') +
    theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
      labels = trans_format("log2", math_format(2^.x)))
  p2 <- ggplot(sge_mean_matrix_nested, aes(sge_mean_matrix_nested[,2],resid_varMean[,2])) +
    geom_point(alpha=0.5) +
    #ylim(-500,500) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    xlab('mean gene expression - individual 2') +
    ylab('variance Resid in gene expression - individual 2') +
    theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
      labels = trans_format("log2", math_format(2^.x)))
  p3 <- ggplot(sge_mean_matrix_nested, aes(sge_mean_matrix_nested[,3],resid_varMean[,3])) +
    geom_point(alpha=0.5) +
    #ylim(-500,500) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    xlab('mean gene expression - individual 3') +
    ylab('variance Resid in gene expression - individual 3') +
    theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
      labels = trans_format("log2", math_format(2^.x)))
  
  plot_grid(p1,p2,p3, ncol = 3)
  ggsave(paste0(figDir,"c",cluster,"/","varResidmean_v_mean_ge__by_individual_c",cluster,".png"))
  
  p1 <- ggplot(sge_mean_matrix_nested, aes(sge_mean_matrix_nested[,1],resid_varMean[,1])) +
    geom_point(alpha=0.5) +
    ylim(-500,500) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    xlab('mean gene expression - individual 1') +
    ylab('variance Resid in gene expression - individual 1') +
    theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
      labels = trans_format("log2", math_format(2^.x)))
  p2 <- ggplot(sge_mean_matrix_nested, aes(sge_mean_matrix_nested[,2],resid_varMean[,2])) +
    geom_point(alpha=0.5) +
    ylim(-500,500) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    xlab('mean gene expression - individual 2') +
    ylab('variance Resid in gene expression - individual 2') +
    theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
      labels = trans_format("log2", math_format(2^.x)))
  p3 <- ggplot(sge_mean_matrix_nested, aes(sge_mean_matrix_nested[,3],resid_varMean[,3])) +
    geom_point(alpha=0.5) +
    ylim(-500,500) +
    #xlim(0,50) +
    stat_smooth(method='lm', color='black', fullrange = T) +
    xlab('mean gene expression - individual 3') +
    ylab('variance Resid in gene expression - individual 3') +
    theme_minimal() + scale_x_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x),
      labels = trans_format("log2", math_format(2^.x)))
  
  plot_grid(p1,p2,p3, ncol = 3)
  ggsave(paste0(figDir,"c",cluster,"/","varResidmean_v_mean_ge__by_individual_yLim500_c",cluster,".png"))
  
  
  ######how do the variance and the DM models compare
  
  res_fullvarDF <- get(paste0("res_fullDF_var_c",cluster))
  res_fullDF <- get(paste0("res_fullDFc",cluster))
  
  
  lpsEloR <- cor.test(res_fullvarDF$beta_trtLPS.elo_centered / (res_fullvarDF$sdev_trtLPS.elo_centered)^0.5,
                      res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5)$estimate
  
  ggplot() +
    geom_point(aes(x = res_fullvarDF$beta_trtLPS.elo_centered / (res_fullvarDF$sdev_trtLPS.elo_centered)^0.5,
                   y = res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5),
               alpha = .1) +
    xlab("variance model - std betas LPS:elo") +
    ylab("DM model - std betas LPS:elo") +
    geom_smooth(aes(x = res_fullvarDF$beta_trtLPS.elo_centered / (res_fullvarDF$sdev_trtLPS.elo_centered)^0.5,
                   y = res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5), method = "lm") +
    geom_text(aes(x = .95 * min(res_fullvarDF$beta_trtLPS.elo_centered / (res_fullvarDF$sdev_trtLPS.elo_centered)^0.5),
                  y = .95 * max(res_fullDF$beta_trtLPS.elo_centered / (res_fullDF$sdev_trtLPS.elo_centered)^0.5)),
                  label = paste0("R = ",round(lpsEloR, digits = 3)))
  ggsave(paste0(figDir,"c",cluster,"/","stdBeta_comp_LPSelo_DM_v_Var_c",cluster,".png"))
  
  
  trtR <- cor.test(res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5,
                   res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5)$estimate
  
  ggplot() +
    geom_point(aes(x = res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5,
                   y = res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5),
               alpha = .1) +
    xlab("variance model - std betas LPS") +
    ylab("DM model - std betas LPS") +
    geom_smooth(aes(x = res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5,
                   y = res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5), method = "lm") +
    geom_text(aes(x = .95 * min(res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5),
                  y = .95 * max(res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5)),
                  label = paste0("R = ",round(trtR, digits = 3)))
  ggsave(paste0(figDir,"c",cluster,"/","stdBeta_comp_Trt_DM_v_Var_c",cluster,".png"))
  
  VARstdBetas <- res_fullvarDF$beta_trtNC / (res_fullvarDF$sdev_trtNC)^0.5
  DMstdBetas <- res_fullDF$beta_trtNC / (res_fullDF$sdev_trtNC)^0.5
  
  ggplot() +
    geom_histogram(aes(x = VARstdBetas), fill = "red", alpha = .4, bins = 50) +
    geom_histogram(aes(x = DMstdBetas), fill = "blue", alpha = .4, bins = 50) +
    xlab("Treatment Standardized Betas (----> Higher Var in NC)") +
    geom_text(aes(x = -20,
                  y = 850,
                  label = "Var Model"), col = "red") +
    geom_text(aes(x = -20,
                  y = 700,
                  label = "DM Model"), col = "blue")
  
  ggsave(paste0(figDir,"c",cluster,"/","stdBeta_hist_Trt_DM_and_Var__c",cluster,".png"))
  
  DMstdBetasSIG <- DMstdBetas[qvalue(p = res_fullDF$p_value_trtNC, fdr.level = .1)$significant]
  VARstdBetasSIG <- VARstdBetas[qvalue(p = res_fullvarDF$p_value_trtNC, fdr.level = .1)$significant]
  
  ggplot() +
    geom_histogram(aes(x = VARstdBetasSIG), fill = "red", alpha = .4, bins = 50) +
    geom_histogram(aes(x = DMstdBetasSIG), fill = "blue", alpha = .4, bins = 50) +
    xlab("Significant Treatment Standardized Betas (----> Higher Var in NC)") +
    geom_text(aes(x = -20,
                  y = 750,
                  label = "Var Model"), col = "red") +
    geom_text(aes(x = -20,
                  y = 700,
                  label = "DM Model"), col = "blue")
  
  ggsave(paste0(figDir,"c",cluster,"/","stdBeta_hist_Trt_DM_and_Var__sigOnly_c",cluster,".png"))
```


